// encodable/lib/options/Encodable.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _globalBox = require("global-box");

var _resolvers = require("./resolvers");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const CACHE_KEY = '@encodable/encodable:options';

class OptionsManager {
  constructor() {
    _defineProperty(this, "state", void 0);

    this.state = (0, _globalBox.getStore)().getOrCreate(CACHE_KEY, () => ({}));
  }

  getNumberFormatResolver() {
    var _this$state$numberFor;

    return (_this$state$numberFor = this.state.numberFormatResolver) != null ? _this$state$numberFor : _resolvers.defaultNumberFormatResolver;
  }

  setNumberFormatResolver(resolver) {
    this.state.numberFormatResolver = resolver;
    return this;
  }

  resetNumberFormatResolver() {
    return this.setNumberFormatResolver(undefined);
  }

  resolveNumberFormat(...params) {
    return this.getNumberFormatResolver()(...params);
  }

  getTimeFormatResolver() {
    var _this$state$timeForma;

    return (_this$state$timeForma = this.state.timeFormatResolver) != null ? _this$state$timeForma : _resolvers.defaultTimeFormatResolver;
  }

  setTimeFormatResolver(resolver) {
    this.state.timeFormatResolver = resolver;
    return this;
  }

  resetTimeFormatResolver() {
    return this.setTimeFormatResolver(undefined);
  }

  resolveTimeFormat(...params) {
    return this.getTimeFormatResolver()(...params);
  }

  getCategoricalColorScaleResolver() {
    var _this$state$categoric;

    return (_this$state$categoric = this.state.categoricalColorScaleResolver) != null ? _this$state$categoric : _resolvers.defaultCategoricalColorScaleResolver;
  }

  setCategoricalColorScaleResolver(resolver) {
    this.state.categoricalColorScaleResolver = resolver;
    return this;
  }

  resetCategoricalColorScaleResolver() {
    return this.setCategoricalColorScaleResolver(undefined);
  }

  resolveCategoricalColorScale(...params) {
    return this.getCategoricalColorScaleResolver()(...params);
  }

  getColorSchemeResolver() {
    var _this$state$colorSche;

    return (_this$state$colorSche = this.state.colorSchemeResolver) != null ? _this$state$colorSche : _resolvers.defaultColorSchemeResolver;
  }

  setColorSchemeResolver(resolver) {
    this.state.colorSchemeResolver = resolver;
    return this;
  }

  resetColorSchemeResolver() {
    return this.setColorSchemeResolver(undefined);
  }

  resolveColorScheme(...params) {
    return this.getColorSchemeResolver()(...params);
  }

}

const Encodable = new OptionsManager();
var _default = Encodable;
exports.default = _default;
