// @superset-ui/core/lib/utils/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  convertKeysToCamelCase: true,
  ensureIsArray: true,
  ensureIsInt: true,
  isDefined: true,
  isRequired: true,
  makeSingleton: true,
  promiseTimeout: true,
  logging: true,
  removeDuplicates: true
};
exports.removeDuplicates = exports.promiseTimeout = exports.makeSingleton = exports.logging = exports.isRequired = exports.isDefined = exports.ensureIsInt = exports.ensureIsArray = exports.convertKeysToCamelCase = void 0;

var _convertKeysToCamelCase = _interopRequireDefault(require("./convertKeysToCamelCase"));

exports.convertKeysToCamelCase = _convertKeysToCamelCase.default;

var _ensureIsArray = _interopRequireDefault(require("./ensureIsArray"));

exports.ensureIsArray = _ensureIsArray.default;

var _ensureIsInt = _interopRequireDefault(require("./ensureIsInt"));

exports.ensureIsInt = _ensureIsInt.default;

var _isDefined = _interopRequireDefault(require("./isDefined"));

exports.isDefined = _isDefined.default;

var _isRequired = _interopRequireDefault(require("./isRequired"));

exports.isRequired = _isRequired.default;

var _makeSingleton = _interopRequireDefault(require("./makeSingleton"));

exports.makeSingleton = _makeSingleton.default;

var _promiseTimeout = _interopRequireDefault(require("./promiseTimeout"));

exports.promiseTimeout = _promiseTimeout.default;

var _logging = _interopRequireDefault(require("./logging"));

exports.logging = _logging.default;

var _removeDuplicates = _interopRequireDefault(require("./removeDuplicates"));

exports.removeDuplicates = _removeDuplicates.default;

var _featureFlags = require("./featureFlags");

Object.keys(_featureFlags).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _featureFlags[key]) return;
  exports[key] = _featureFlags[key];
});

var _random = require("./random");

Object.keys(_random).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _random[key]) return;
  exports[key] = _random[key];
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
