// @loaders.gl/gltf/dist/es5/lib/extensions/gltf-extensions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decodeExtensions = decodeExtensions;
exports.EXTENSIONS = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var KHR_draco_mesh_compression = _interopRequireWildcard(require("./KHR_draco_mesh_compression"));

var KHR_materials_unlit = _interopRequireWildcard(require("./KHR_materials_unlit"));

var KHR_lights_punctual = _interopRequireWildcard(require("./KHR_lights_punctual"));

var KHR_techniques_webgl = _interopRequireWildcard(require("./KHR_techniques_webgl"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var EXTENSIONS = {
  KHR_draco_mesh_compression: KHR_draco_mesh_compression,
  KHR_materials_unlit: KHR_materials_unlit,
  KHR_lights_punctual: KHR_lights_punctual,
  KHR_techniques_webgl: KHR_techniques_webgl
};
exports.EXTENSIONS = EXTENSIONS;

function decodeExtensions(_x) {
  return _decodeExtensions.apply(this, arguments);
}

function _decodeExtensions() {
  _decodeExtensions = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(gltf) {
    var options,
        context,
        _extensionName,
        _options$gltf,
        excludes,
        exclude,
        extension,
        _args = arguments;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            options = _args.length > 1 && _args[1] !== undefined ? _args[1] : {};
            context = _args.length > 2 ? _args[2] : undefined;
            _context.t0 = _regenerator.default.keys(EXTENSIONS);

          case 3:
            if ((_context.t1 = _context.t0()).done) {
              _context.next = 13;
              break;
            }

            _extensionName = _context.t1.value;
            excludes = (options === null || options === void 0 ? void 0 : (_options$gltf = options.gltf) === null || _options$gltf === void 0 ? void 0 : _options$gltf.excludeExtensions) || {};
            exclude = _extensionName in excludes && !excludes[_extensionName];

            if (exclude) {
              _context.next = 11;
              break;
            }

            extension = EXTENSIONS[_extensionName];
            _context.next = 11;
            return extension.decode(gltf, options, context);

          case 11:
            _context.next = 3;
            break;

          case 13:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _decodeExtensions.apply(this, arguments);
}
