// @superset-ui/core/lib/time-format/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  TimeFormats: true,
  TimeFormatter: true,
  getTimeFormatterRegistry: true,
  createD3TimeFormatter: true,
  createMultiFormatter: true,
  smartDateFormatter: true,
  smartDateDetailedFormatter: true,
  smartDateVerboseFormatter: true,
  LOCAL_PREFIX: true,
  PREVIEW_TIME: true,
  formatTime: true,
  formatTimeRange: true,
  getTimeFormatter: true,
  getTimeFormatterForGranularity: true,
  getTimeRangeFormatter: true
};
exports.smartDateVerboseFormatter = exports.smartDateFormatter = exports.smartDateDetailedFormatter = exports.getTimeRangeFormatter = exports.getTimeFormatterRegistry = exports.getTimeFormatterForGranularity = exports.getTimeFormatter = exports.formatTimeRange = exports.formatTime = exports.createMultiFormatter = exports.createD3TimeFormatter = exports.TimeFormatter = exports.TimeFormats = exports.PREVIEW_TIME = exports.LOCAL_PREFIX = void 0;

var _TimeFormats = _interopRequireWildcard(require("./TimeFormats"));

exports.TimeFormats = _TimeFormats.default;
exports.LOCAL_PREFIX = _TimeFormats.LOCAL_PREFIX;

var _TimeFormatter = _interopRequireWildcard(require("./TimeFormatter"));

exports.TimeFormatter = _TimeFormatter.default;
exports.PREVIEW_TIME = _TimeFormatter.PREVIEW_TIME;

var _TimeFormatterRegistrySingleton = _interopRequireWildcard(require("./TimeFormatterRegistrySingleton"));

exports.getTimeFormatterRegistry = _TimeFormatterRegistrySingleton.default;
exports.formatTime = _TimeFormatterRegistrySingleton.formatTime;
exports.formatTimeRange = _TimeFormatterRegistrySingleton.formatTimeRange;
exports.getTimeFormatter = _TimeFormatterRegistrySingleton.getTimeFormatter;
exports.getTimeFormatterForGranularity = _TimeFormatterRegistrySingleton.getTimeFormatterForGranularity;
exports.getTimeRangeFormatter = _TimeFormatterRegistrySingleton.getTimeRangeFormatter;

var _createD3TimeFormatter = _interopRequireDefault(require("./factories/createD3TimeFormatter"));

exports.createD3TimeFormatter = _createD3TimeFormatter.default;

var _createMultiFormatter = _interopRequireDefault(require("./factories/createMultiFormatter"));

exports.createMultiFormatter = _createMultiFormatter.default;

var _smartDate = _interopRequireDefault(require("./formatters/smartDate"));

exports.smartDateFormatter = _smartDate.default;

var _smartDateDetailed = _interopRequireDefault(require("./formatters/smartDateDetailed"));

exports.smartDateDetailedFormatter = _smartDateDetailed.default;

var _smartDateVerbose = _interopRequireDefault(require("./formatters/smartDateVerbose"));

exports.smartDateVerboseFormatter = _smartDateVerbose.default;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  exports[key] = _types[key];
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
