// @superset-ui/core/lib/time-format/TimeFormatter.js
"use strict";

exports.__esModule = true;
exports.default = exports.PREVIEW_TIME = void 0;

var _models = require("../models");

var _utils = require("../utils");

var _stringifyTimeInput = _interopRequireDefault(require("./utils/stringifyTimeInput"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const PREVIEW_TIME = new Date(Date.UTC(2017, 1, 14, 11, 22, 33)); // Use type augmentation to indicate that
// an instance of TimeFormatter is also a function

exports.PREVIEW_TIME = PREVIEW_TIME;

class TimeFormatter extends _models.ExtensibleFunction {
  constructor(config) {
    super(value => this.format(value));
    this.id = void 0;
    this.label = void 0;
    this.description = void 0;
    this.formatFunc = void 0;
    this.useLocalTime = void 0;
    const {
      id = (0, _utils.isRequired)('config.id'),
      label,
      description = '',
      formatFunc = (0, _utils.isRequired)('config.formatFunc'),
      useLocalTime = false
    } = config;
    this.id = id;
    this.label = label ?? id;
    this.description = description;
    this.formatFunc = formatFunc;
    this.useLocalTime = useLocalTime;
  }

  format(value) {
    return (0, _stringifyTimeInput.default)(value, time => this.formatFunc(time));
  }

  preview(value = PREVIEW_TIME) {
    return `${value.toUTCString()} => ${this.format(value)}`;
  }

}

var _default = TimeFormatter;
exports.default = _default;
