// @superset-ui/core/lib/time-format/TimeFormatterRegistrySingleton.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
exports.formatTime = formatTime;
exports.formatTimeRange = formatTimeRange;
exports.getTimeFormatter = getTimeFormatter;
exports.getTimeFormatterForGranularity = getTimeFormatterForGranularity;
exports.getTimeRangeFormatter = getTimeRangeFormatter;

var _utils = require("../utils");

var _TimeFormatterRegistry = _interopRequireDefault(require("./TimeFormatterRegistry"));

var _TimeFormatter = _interopRequireDefault(require("./TimeFormatter"));

var _TimeFormatsForGranularity = _interopRequireDefault(require("./TimeFormatsForGranularity"));

var _TimeFormats = require("./TimeFormats");

var _createTimeRangeFromGranularity = _interopRequireDefault(require("./utils/createTimeRangeFromGranularity"));

var _TimeRangeFormatter = _interopRequireDefault(require("./TimeRangeFormatter"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const getInstance = (0, _utils.makeSingleton)(_TimeFormatterRegistry.default);
var _default = getInstance;
exports.default = _default;

function getTimeRangeFormatter(formatId) {
  return new _TimeRangeFormatter.default({
    id: formatId || 'undefined',
    formatFunc: range => {
      const format = getInstance().get(formatId);
      const [start, end] = range.map(value => format(value));
      return start === end ? start : [start, end].join(' — ');
    },
    useLocalTime: formatId == null ? void 0 : formatId.startsWith(_TimeFormats.LOCAL_PREFIX)
  });
}

function formatTimeRange(formatId, range) {
  return getTimeRangeFormatter(formatId)(range);
}

function getTimeFormatter(formatId, granularity) {
  if (granularity) {
    const formatString = formatId || _TimeFormatsForGranularity.default[granularity];
    const timeRangeFormatter = getTimeRangeFormatter(formatString);
    return new _TimeFormatter.default({
      id: [formatString, granularity].join('/'),
      formatFunc: value => timeRangeFormatter.format((0, _createTimeRangeFromGranularity.default)(value, granularity, timeRangeFormatter.useLocalTime)),
      useLocalTime: timeRangeFormatter.useLocalTime
    });
  }

  return getInstance().get(formatId);
}
/**
 * Syntactic sugar for backward compatibility
 * TODO: Deprecate this in the next breaking change.
 * @param granularity
 */


function getTimeFormatterForGranularity(granularity) {
  return getTimeFormatter(undefined, granularity);
}

function formatTime(formatId, value, granularity) {
  return getTimeFormatter(formatId, granularity)(value);
}
