// @superset-ui/legacy-plugin-chart-force-directed/lib/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var _default = {
  controlPanelSections: [_chartControls.sections.legacyRegularTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [['groupby'], ['metric'], ['adhoc_filters'], ['row_limit'], [{
      name: 'sort_by_metric',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Sort by metric'),
        description: (0, _core.t)('Whether to sort results by the selected metric in descending order.')
      }
    }]]
  }, {
    label: (0, _core.t)('Chart Options'),
    expanded: true,
    controlSetRows: [[{
      name: 'link_length',
      config: {
        type: 'SelectControl',
        renderTrigger: true,
        freeForm: true,
        label: (0, _core.t)('Link Length'),
        default: '200',
        choices: (0, _chartControls.formatSelectOptions)(['10', '25', '50', '75', '100', '150', '200', '250']),
        description: (0, _core.t)('Link length in the force layout')
      }
    }], [{
      name: 'charge',
      config: {
        type: 'SelectControl',
        renderTrigger: true,
        freeForm: true,
        label: (0, _core.t)('Charge'),
        default: '-500',
        choices: (0, _chartControls.formatSelectOptions)(['-50', '-75', '-100', '-150', '-200', '-250', '-500', '-1000', '-2500', '-5000']),
        description: (0, _core.t)('Charge in the force layout')
      }
    }]]
  }],
  controlOverrides: {
    groupby: {
      label: (0, _core.t)('Source / Target'),
      description: (0, _core.t)('Choose a source and a target')
    }
  }
};
exports.default = _default;
