// @superset-ui/plugin-chart-echarts/lib/MixedTimeseries/EchartsMixedTimeseries.js
"use strict";

exports.__esModule = true;
exports.default = EchartsMixedTimeseries;

var _react = _interopRequireWildcard(require("react"));

var _Echart = _interopRequireDefault(require("../components/Echart"));

var _series = require("../utils/series");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
function EchartsMixedTimeseries({
  height,
  width,
  echartOptions,
  setDataMask,
  labelMap,
  labelMapB,
  groupby,
  groupbyB,
  selectedValues,
  formData,
  seriesBreakdown
}) {
  const isFirstQuery = (0, _react.useCallback)(seriesIndex => seriesIndex < seriesBreakdown, [seriesBreakdown]);
  const handleChange = (0, _react.useCallback)((values, seriesIndex) => {
    const emitFilter = isFirstQuery(seriesIndex) ? formData.emitFilter : formData.emitFilterB;

    if (!emitFilter) {
      return;
    }

    const currentGroupBy = isFirstQuery(seriesIndex) ? groupby : groupbyB;
    const currentLabelMap = isFirstQuery(seriesIndex) ? labelMap : labelMapB;
    const groupbyValues = values.map(value => currentLabelMap[value]).filter(value => !!value);
    setDataMask({
      extraFormData: {
        // @ts-ignore
        filters: values.length === 0 ? [] : [...currentGroupBy.map((col, idx) => {
          const val = groupbyValues.map(v => v[idx]);
          if (val === null || val === undefined) return {
            col,
            op: 'IS NULL'
          };
          return {
            col,
            op: 'IN',
            val: val
          };
        })]
      },
      filterState: {
        value: !groupbyValues.length ? null : groupbyValues,
        selectedValues: values.length ? values : null
      }
    });
  }, [groupby, groupbyB, labelMap, labelMapB, setDataMask, selectedValues]);
  const eventHandlers = {
    click: props => {
      const {
        seriesName,
        seriesIndex
      } = props;
      const values = Object.values(selectedValues);

      if (values.includes(seriesName)) {
        handleChange(values.filter(v => v !== seriesName), seriesIndex);
      } else {
        handleChange([seriesName], seriesIndex);
      }
    },
    mouseout: () => {
      _series.currentSeries.name = '';
    },
    mouseover: params => {
      _series.currentSeries.name = params.seriesName;
    }
  };
  return (0, _react2.jsx)(_Echart.default, {
    height: height,
    width: width,
    echartOptions: echartOptions,
    eventHandlers: eventHandlers,
    selectedValues: selectedValues
  });
}
