// @superset-ui/plugin-chart-echarts/lib/MixedTimeseries/transformProps.js
"use strict";

exports.__esModule = true;
exports.default = transformProps;

var _core = require("@superset-ui/core");

var _types = require("./types");

var _types2 = require("../types");

var _controls = require("../utils/controls");

var _series = require("../utils/series");

var _annotation = require("../utils/annotation");

var _prophet = require("../utils/prophet");

var _defaults = require("../defaults");

var _transformers = require("../Timeseries/transformers");

var _constants = require("../constants");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable camelcase */
function transformProps(chartProps) {
  const {
    width,
    height,
    formData,
    queriesData,
    hooks,
    filterState
  } = chartProps;
  const {
    annotation_data: annotationData_
  } = queriesData[0];
  const annotationData = annotationData_ || {};
  const data1 = queriesData[0].data || [];
  const data2 = queriesData[1].data || [];
  const {
    area,
    areaB,
    annotationLayers,
    colorScheme,
    contributionMode,
    legendOrientation,
    legendType,
    logAxis,
    logAxisSecondary,
    markerEnabled,
    markerEnabledB,
    markerSize,
    markerSizeB,
    opacity,
    opacityB,
    minorSplitLine,
    seriesType,
    seriesTypeB,
    showLegend,
    showValue,
    showValueB,
    stack,
    stackB,
    truncateYAxis,
    tooltipTimeFormat,
    yAxisFormat,
    yAxisFormatSecondary,
    xAxisTimeFormat,
    yAxisBounds,
    yAxisIndex,
    yAxisIndexB,
    yAxisTitleSecondary,
    zoomable,
    richTooltip,
    tooltipSortByMetric,
    xAxisLabelRotation,
    groupby,
    groupbyB,
    emitFilter,
    emitFilterB,
    xAxisTitle,
    yAxisTitle,
    xAxisTitleMargin,
    yAxisTitleMargin,
    yAxisTitlePosition
  } = { ..._types.DEFAULT_FORM_DATA,
    ...formData
  };

  const colorScale = _core.CategoricalColorNamespace.getScale(colorScheme);

  const rawSeriesA = (0, _series.extractTimeseriesSeries)((0, _prophet.rebaseTimeseriesDatum)(data1), {
    fillNeighborValue: stack ? 0 : undefined
  });
  const rawSeriesB = (0, _series.extractTimeseriesSeries)((0, _prophet.rebaseTimeseriesDatum)(data2), {
    fillNeighborValue: stackB ? 0 : undefined
  });
  const series = [];
  const formatter = (0, _core.getNumberFormatter)(contributionMode ? ',.0%' : yAxisFormat);
  const formatterSecondary = (0, _core.getNumberFormatter)(contributionMode ? ',.0%' : yAxisFormatSecondary);
  const primarySeries = new Set();
  const secondarySeries = new Set();

  const mapSeriesIdToAxis = (seriesOption, index) => {
    if (index === 1) {
      secondarySeries.add(seriesOption.id);
    } else {
      primarySeries.add(seriesOption.id);
    }
  };

  rawSeriesA.forEach(seriesOption => mapSeriesIdToAxis(seriesOption, yAxisIndex));
  rawSeriesB.forEach(seriesOption => mapSeriesIdToAxis(seriesOption, yAxisIndexB));
  rawSeriesA.forEach(entry => {
    const transformedSeries = (0, _transformers.transformSeries)(entry, colorScale, {
      area,
      markerEnabled,
      markerSize,
      areaOpacity: opacity,
      seriesType,
      showValue,
      stack,
      yAxisIndex,
      filterState
    });
    if (transformedSeries) series.push(transformedSeries);
  });
  rawSeriesB.forEach(entry => {
    const transformedSeries = (0, _transformers.transformSeries)(entry, colorScale, {
      area: areaB,
      markerEnabled: markerEnabledB,
      markerSize: markerSizeB,
      areaOpacity: opacityB,
      seriesType: seriesTypeB,
      showValue: showValueB,
      stack: stackB,
      yAxisIndex: yAxisIndexB,
      filterState
    });
    if (transformedSeries) series.push(transformedSeries);
  });
  annotationLayers.filter(layer => layer.show).forEach(layer => {
    if ((0, _core.isFormulaAnnotationLayer)(layer)) series.push((0, _transformers.transformFormulaAnnotation)(layer, data1, colorScale));else if ((0, _core.isIntervalAnnotationLayer)(layer)) {
      series.push(...(0, _transformers.transformIntervalAnnotation)(layer, data1, annotationData, colorScale));
    } else if ((0, _core.isEventAnnotationLayer)(layer)) {
      series.push(...(0, _transformers.transformEventAnnotation)(layer, data1, annotationData, colorScale));
    } else if ((0, _core.isTimeseriesAnnotationLayer)(layer)) {
      series.push(...(0, _transformers.transformTimeseriesAnnotation)(layer, markerSize, data1, annotationData));
    }
  }); // yAxisBounds need to be parsed to replace incompatible values with undefined

  let [min, max] = (yAxisBounds || []).map(_controls.parseYAxisBound); // default to 0-100% range when doing row-level contribution chart

  if (contributionMode === 'row' && stack) {
    if (min === undefined) min = 0;
    if (max === undefined) max = 1;
  }

  const tooltipTimeFormatter = (0, _transformers.getTooltipTimeFormatter)(tooltipTimeFormat);
  const xAxisFormatter = (0, _transformers.getXAxisFormatter)(xAxisTimeFormat);
  const addYAxisTitleOffset = !!(yAxisTitle || yAxisTitleSecondary);
  const addXAxisTitleOffset = !!xAxisTitle;
  const chartPadding = (0, _transformers.getPadding)(showLegend, legendOrientation, addYAxisTitleOffset, zoomable, null, addXAxisTitleOffset, yAxisTitlePosition, yAxisTitleMargin, xAxisTitleMargin);
  const labelMap = rawSeriesA.reduce((acc, datum) => {
    const label = datum.name;
    return { ...acc,
      [label]: label.split(', ')
    };
  }, {});
  const labelMapB = rawSeriesB.reduce((acc, datum) => {
    const label = datum.name;
    return { ...acc,
      [label]: label.split(', ')
    };
  }, {});
  const {
    setDataMask = () => {}
  } = hooks;
  const echartOptions = {
    useUTC: true,
    grid: { ..._defaults.defaultGrid,
      ...chartPadding
    },
    xAxis: {
      type: 'time',
      name: xAxisTitle,
      nameGap: xAxisTitleMargin,
      nameLocation: 'middle',
      axisLabel: {
        formatter: xAxisFormatter,
        rotate: xAxisLabelRotation
      }
    },
    yAxis: [{ ..._defaults.defaultYAxis,
      type: logAxis ? 'log' : 'value',
      min,
      max,
      minorTick: {
        show: true
      },
      minorSplitLine: {
        show: minorSplitLine
      },
      axisLabel: {
        formatter
      },
      scale: truncateYAxis,
      name: yAxisTitle,
      nameGap: yAxisTitleMargin,
      nameLocation: yAxisTitlePosition === 'Left' ? 'middle' : 'end'
    }, { ..._defaults.defaultYAxis,
      type: logAxisSecondary ? 'log' : 'value',
      min,
      max,
      minorTick: {
        show: true
      },
      splitLine: {
        show: false
      },
      minorSplitLine: {
        show: minorSplitLine
      },
      axisLabel: {
        formatter: formatterSecondary
      },
      scale: truncateYAxis,
      name: yAxisTitleSecondary
    }],
    tooltip: { ..._defaults.defaultTooltip,
      appendToBody: true,
      trigger: richTooltip ? 'axis' : 'item',
      formatter: params => {
        const xValue = richTooltip ? params[0].value[0] : params.value[0];
        const prophetValue = richTooltip ? params : [params];

        if (richTooltip && tooltipSortByMetric) {
          prophetValue.sort((a, b) => b.data[1] - a.data[1]);
        }

        const rows = [`${tooltipTimeFormatter(xValue)}`];
        const prophetValues = (0, _prophet.extractProphetValuesFromTooltipParams)(prophetValue);
        Object.keys(prophetValues).forEach(key => {
          const value = prophetValues[key];
          const content = (0, _prophet.formatProphetTooltipSeries)({ ...value,
            seriesName: key,
            formatter: primarySeries.has(key) ? formatter : formatterSecondary
          });

          if (_series.currentSeries.name === key) {
            rows.push(`<span style="font-weight: 700">${content}</span>`);
          } else {
            rows.push(`<span style="opacity: 0.7">${content}</span>`);
          }
        });
        return rows.join('<br />');
      }
    },
    legend: { ...(0, _series.getLegendProps)(legendType, legendOrientation, showLegend, zoomable),
      // @ts-ignore
      data: rawSeriesA.concat(rawSeriesB).filter(entry => (0, _prophet.extractForecastSeriesContext)(entry.name || '').type === _types2.ForecastSeriesEnum.Observation).map(entry => entry.name || '').concat((0, _annotation.extractAnnotationLabels)(annotationLayers, annotationData))
    },
    series: (0, _series.dedupSeries)(series),
    toolbox: {
      show: zoomable,
      top: _constants.TIMESERIES_CONSTANTS.toolboxTop,
      right: _constants.TIMESERIES_CONSTANTS.toolboxRight,
      feature: {
        dataZoom: {
          yAxisIndex: false,
          title: {
            zoom: 'zoom area',
            back: 'restore zoom'
          }
        }
      }
    },
    dataZoom: zoomable ? [{
      type: 'slider',
      start: _constants.TIMESERIES_CONSTANTS.dataZoomStart,
      end: _constants.TIMESERIES_CONSTANTS.dataZoomEnd,
      bottom: _constants.TIMESERIES_CONSTANTS.zoomBottom
    }] : []
  };
  return {
    formData,
    width,
    height,
    echartOptions,
    setDataMask,
    emitFilter,
    emitFilterB,
    labelMap,
    labelMapB,
    groupby,
    groupbyB,
    seriesBreakdown: rawSeriesA.length,
    selectedValues: filterState.selectedValues || []
  };
}
