// @superset-ui/plugin-chart-echarts/lib/MixedTimeseries/types.js
"use strict";

exports.__esModule = true;
exports.DEFAULT_FORM_DATA = void 0;

var _types = require("../types");

var _types2 = require("../Timeseries/types");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// @ts-ignore
const DEFAULT_FORM_DATA = { ..._types.DEFAULT_LEGEND_FORM_DATA,
  annotationLayers: [],
  minorSplitLine: _types2.DEFAULT_FORM_DATA.minorSplitLine,
  truncateYAxis: _types2.DEFAULT_FORM_DATA.truncateYAxis,
  truncateYAxisSecondary: _types2.DEFAULT_FORM_DATA.truncateYAxis,
  logAxis: _types2.DEFAULT_FORM_DATA.logAxis,
  logAxisSecondary: _types2.DEFAULT_FORM_DATA.logAxis,
  yAxisBounds: _types2.DEFAULT_FORM_DATA.yAxisBounds,
  yAxisBoundsSecondary: _types2.DEFAULT_FORM_DATA.yAxisBounds,
  yAxisFormat: _types2.DEFAULT_FORM_DATA.yAxisFormat,
  yAxisFormatSecondary: _types2.DEFAULT_FORM_DATA.yAxisFormat,
  yAxisTitleSecondary: _types.DEFAULT_TITLE_FORM_DATA.yAxisTitle,
  tooltipTimeFormat: _types2.DEFAULT_FORM_DATA.tooltipTimeFormat,
  xAxisTimeFormat: _types2.DEFAULT_FORM_DATA.xAxisTimeFormat,
  area: _types2.DEFAULT_FORM_DATA.area,
  areaB: _types2.DEFAULT_FORM_DATA.area,
  markerEnabled: _types2.DEFAULT_FORM_DATA.markerEnabled,
  markerEnabledB: _types2.DEFAULT_FORM_DATA.markerEnabled,
  markerSize: _types2.DEFAULT_FORM_DATA.markerSize,
  markerSizeB: _types2.DEFAULT_FORM_DATA.markerSize,
  opacity: _types2.DEFAULT_FORM_DATA.opacity,
  opacityB: _types2.DEFAULT_FORM_DATA.opacity,
  orderDesc: _types2.DEFAULT_FORM_DATA.orderDesc,
  orderDescB: _types2.DEFAULT_FORM_DATA.orderDesc,
  rowLimit: _types2.DEFAULT_FORM_DATA.rowLimit,
  rowLimitB: _types2.DEFAULT_FORM_DATA.rowLimit,
  seriesType: _types2.DEFAULT_FORM_DATA.seriesType,
  seriesTypeB: _types2.DEFAULT_FORM_DATA.seriesType,
  showValue: _types2.DEFAULT_FORM_DATA.showValue,
  showValueB: _types2.DEFAULT_FORM_DATA.showValue,
  stack: _types2.DEFAULT_FORM_DATA.stack,
  stackB: _types2.DEFAULT_FORM_DATA.stack,
  yAxisIndex: 0,
  yAxisIndexB: 0,
  groupby: [],
  groupbyB: [],
  zoomable: _types2.DEFAULT_FORM_DATA.zoomable,
  richTooltip: _types2.DEFAULT_FORM_DATA.richTooltip,
  xAxisLabelRotation: _types2.DEFAULT_FORM_DATA.xAxisLabelRotation,
  ..._types.DEFAULT_TITLE_FORM_DATA
};
exports.DEFAULT_FORM_DATA = DEFAULT_FORM_DATA;
