// @superset-ui/plugin-chart-echarts/lib/Pie/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

var _types = require("./types");

var _controls = require("../controls");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const {
  donut,
  innerRadius,
  labelsOutside,
  labelType,
  labelLine,
  outerRadius,
  numberFormat,
  showLabels
} = _types.DEFAULT_FORM_DATA;
const config = {
  controlPanelSections: [_chartControls.sections.legacyRegularTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [['groupby'], ['metric'], ['adhoc_filters'], _chartControls.emitFilterControl, ['row_limit'], [{
      name: 'sort_by_metric',
      config: {
        default: true,
        type: 'CheckboxControl',
        label: (0, _core.t)('Sort by metric'),
        description: (0, _core.t)('Whether to sort results by the selected metric in descending order.')
      }
    }]]
  }, {
    label: (0, _core.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_scheme'], [{
      name: 'show_labels_threshold',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Percentage threshold'),
        renderTrigger: true,
        isFloat: true,
        default: 5,
        description: (0, _core.t)('Minimum threshold in percentage points for showing labels.')
      }
    }], ..._controls.legendSection, // eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Labels'))], [{
      name: 'label_type',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Label Type'),
        default: labelType,
        renderTrigger: true,
        choices: [['key', 'Category Name'], ['value', 'Value'], ['percent', 'Percentage'], ['key_value', 'Category and Value'], ['key_percent', 'Category and Percentage'], ['key_value_percent', 'Category, Value and Percentage']],
        description: (0, _core.t)('What should be shown on the label?')
      }
    }], [{
      name: 'number_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Number format'),
        renderTrigger: true,
        default: numberFormat,
        choices: _chartControls.D3_FORMAT_OPTIONS,
        description: `${(0, _core.t)('D3 format syntax: https://github.com/d3/d3-format')} ${(0, _core.t)('Only applies when "Label Type" is set to show values.')}`
      }
    }], [{
      name: 'date_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Date format'),
        renderTrigger: true,
        choices: _chartControls.D3_TIME_FORMAT_OPTIONS,
        default: 'smart_date',
        description: _chartControls.D3_FORMAT_DOCS
      }
    }], [{
      name: 'show_labels',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Show Labels'),
        renderTrigger: true,
        default: showLabels,
        description: (0, _core.t)('Whether to display the labels.')
      }
    }], [{
      name: 'labels_outside',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Put labels outside'),
        default: labelsOutside,
        renderTrigger: true,
        description: (0, _core.t)('Put the labels outside of the pie?'),
        visibility: ({
          controls
        }) => {
          var _controls$show_labels;

          return Boolean(controls == null ? void 0 : (_controls$show_labels = controls.show_labels) == null ? void 0 : _controls$show_labels.value);
        }
      }
    }], [{
      name: 'label_line',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Label Line'),
        default: labelLine,
        renderTrigger: true,
        description: (0, _core.t)('Draw line from Pie to label when labels outside?'),
        visibility: ({
          controls
        }) => {
          var _controls$show_labels2;

          return Boolean(controls == null ? void 0 : (_controls$show_labels2 = controls.show_labels) == null ? void 0 : _controls$show_labels2.value);
        }
      }
    }], // eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Pie shape'))], [{
      name: 'outerRadius',
      config: {
        type: 'SliderControl',
        label: (0, _core.t)('Outer Radius'),
        renderTrigger: true,
        min: 10,
        max: 100,
        step: 1,
        default: outerRadius,
        description: (0, _core.t)('Outer edge of Pie chart')
      }
    }], [{
      name: 'donut',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Donut'),
        default: donut,
        renderTrigger: true,
        description: (0, _core.t)('Do you want a donut or a pie?')
      }
    }], [{
      name: 'innerRadius',
      config: {
        type: 'SliderControl',
        label: (0, _core.t)('Inner Radius'),
        renderTrigger: true,
        min: 0,
        max: 100,
        step: 1,
        default: innerRadius,
        description: (0, _core.t)('Inner radius of donut hole'),
        visibility: ({
          controls
        }) => {
          var _controls$donut;

          return Boolean(controls == null ? void 0 : (_controls$donut = controls.donut) == null ? void 0 : _controls$donut.value);
        }
      }
    }]]
  }],
  controlOverrides: {
    series: {
      validators: [_core.validateNonEmpty],
      clearable: false
    },
    row_limit: {
      default: 100
    }
  }
};
var _default = config;
exports.default = _default;
