// @superset-ui/plugin-chart-echarts/lib/Pie/types.js
"use strict";

exports.__esModule = true;
exports.EchartsPieLabelType = exports.DEFAULT_FORM_DATA = void 0;

var _types = require("../types");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
let EchartsPieLabelType;
exports.EchartsPieLabelType = EchartsPieLabelType;

(function (EchartsPieLabelType) {
  EchartsPieLabelType["Key"] = "key";
  EchartsPieLabelType["Value"] = "value";
  EchartsPieLabelType["Percent"] = "percent";
  EchartsPieLabelType["KeyValue"] = "key_value";
  EchartsPieLabelType["KeyPercent"] = "key_percent";
  EchartsPieLabelType["KeyValuePercent"] = "key_value_percent";
})(EchartsPieLabelType || (exports.EchartsPieLabelType = EchartsPieLabelType = {}));

// @ts-ignore
const DEFAULT_FORM_DATA = { ..._types.DEFAULT_LEGEND_FORM_DATA,
  donut: false,
  groupby: [],
  innerRadius: 30,
  labelLine: false,
  labelType: EchartsPieLabelType.Key,
  legendOrientation: _types.LegendOrientation.Top,
  legendType: _types.LegendType.Scroll,
  numberFormat: 'SMART_NUMBER',
  outerRadius: 70,
  showLabels: true,
  labelsOutside: true,
  showLabelsThreshold: 5,
  emitFilter: false,
  dateFormat: 'smart_date'
};
exports.DEFAULT_FORM_DATA = DEFAULT_FORM_DATA;
