// @superset-ui/plugin-chart-echarts/esm/Timeseries/types.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { sections } from '@superset-ui/chart-controls';
import { DEFAULT_LEGEND_FORM_DATA, DEFAULT_TITLE_FORM_DATA } from '../types';
export let EchartsTimeseriesContributionType;

(function (EchartsTimeseriesContributionType) {
  EchartsTimeseriesContributionType["Row"] = "row";
  EchartsTimeseriesContributionType["Column"] = "column";
})(EchartsTimeseriesContributionType || (EchartsTimeseriesContributionType = {}));

export let EchartsTimeseriesSeriesType;

(function (EchartsTimeseriesSeriesType) {
  EchartsTimeseriesSeriesType["Line"] = "line";
  EchartsTimeseriesSeriesType["Scatter"] = "scatter";
  EchartsTimeseriesSeriesType["Smooth"] = "smooth";
  EchartsTimeseriesSeriesType["Bar"] = "bar";
  EchartsTimeseriesSeriesType["Start"] = "start";
  EchartsTimeseriesSeriesType["Middle"] = "middle";
  EchartsTimeseriesSeriesType["End"] = "end";
})(EchartsTimeseriesSeriesType || (EchartsTimeseriesSeriesType = {}));

// @ts-ignore
export const DEFAULT_FORM_DATA = { ...DEFAULT_LEGEND_FORM_DATA,
  annotationLayers: sections.annotationLayers,
  area: false,
  forecastEnabled: sections.FORECAST_DEFAULT_DATA.forecastEnabled,
  forecastInterval: sections.FORECAST_DEFAULT_DATA.forecastInterval,
  forecastPeriods: sections.FORECAST_DEFAULT_DATA.forecastPeriods,
  forecastSeasonalityDaily: sections.FORECAST_DEFAULT_DATA.forecastSeasonalityDaily,
  forecastSeasonalityWeekly: sections.FORECAST_DEFAULT_DATA.forecastSeasonalityWeekly,
  forecastSeasonalityYearly: sections.FORECAST_DEFAULT_DATA.forecastSeasonalityYearly,
  logAxis: false,
  markerEnabled: false,
  markerSize: 6,
  minorSplitLine: false,
  opacity: 0.2,
  orderDesc: true,
  rowLimit: 10000,
  seriesType: EchartsTimeseriesSeriesType.Line,
  stack: false,
  tooltipTimeFormat: 'smart_date',
  truncateYAxis: true,
  yAxisBounds: [null, null],
  zoomable: false,
  richTooltip: true,
  xAxisLabelRotation: 0,
  emitFilter: false,
  groupby: [],
  showValue: false,
  onlyTotal: false,
  ...DEFAULT_TITLE_FORM_DATA
};
