// @superset-ui/plugin-chart-echarts/lib/Radar/transformProps.js
"use strict";

exports.__esModule = true;
exports.default = transformProps;
exports.formatLabel = formatLabel;

var _core = require("@superset-ui/core");

var _types = require("./types");

var _types2 = require("../types");

var _series = require("../utils/series");

var _defaults = require("../defaults");

var _constants = require("../constants");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
function formatLabel({
  params,
  labelType,
  numberFormatter
}) {
  const {
    name = '',
    value
  } = params;
  const formattedValue = numberFormatter(value);

  switch (labelType) {
    case _types.EchartsRadarLabelType.Value:
      return formattedValue;

    case _types.EchartsRadarLabelType.KeyValue:
      return `${name}: ${formattedValue}`;

    default:
      return name;
  }
}

function transformProps(chartProps) {
  const {
    formData,
    height,
    hooks,
    filterState,
    queriesData,
    width
  } = chartProps;
  const {
    data = []
  } = queriesData[0];
  const coltypeMapping = (0, _series.getColtypesMapping)(queriesData[0]);
  const {
    colorScheme,
    groupby,
    labelType,
    labelPosition,
    legendOrientation,
    legendType,
    metrics = [],
    numberFormat,
    dateFormat,
    showLabels,
    showLegend,
    isCircle,
    columnConfig
  } = { ..._types2.DEFAULT_LEGEND_FORM_DATA,
    ..._types.DEFAULT_FORM_DATA,
    ...formData
  };
  const {
    setDataMask = () => {}
  } = hooks;

  const colorFn = _core.CategoricalColorNamespace.getScale(colorScheme);

  const numberFormatter = (0, _core.getNumberFormatter)(numberFormat);

  const formatter = params => formatLabel({
    params,
    numberFormatter,
    labelType
  });

  const metricsLabel = metrics.map(metric => (0, _core.getMetricLabel)(metric));
  const metricLabelAndMaxValueMap = new Map();
  const columnsLabelMap = new Map();
  const transformedData = [];
  data.forEach(datum => {
    const joinedName = (0, _series.extractGroupbyLabel)({
      datum,
      groupby,
      coltypeMapping,
      timeFormatter: (0, _core.getTimeFormatter)(dateFormat)
    }); // map(joined_name: [columnLabel_1, columnLabel_2, ...])

    columnsLabelMap.set(joinedName, groupby.map(col => datum[col])); // put max value of series into metricLabelAndMaxValueMap
    // eslint-disable-next-line no-restricted-syntax

    for (const [metricLabel, value] of Object.entries(datum)) {
      if (metricLabelAndMaxValueMap.has(metricLabel)) {
        metricLabelAndMaxValueMap.set(metricLabel, Math.max(value, (0, _core.ensureIsInt)(metricLabelAndMaxValueMap.get(metricLabel), Number.MIN_SAFE_INTEGER)));
      } else {
        metricLabelAndMaxValueMap.set(metricLabel, value);
      }
    }

    const isFiltered = filterState.selectedValues && !filterState.selectedValues.includes(joinedName); // generate transformedData

    transformedData.push({
      value: metricsLabel.map(metricLabel => datum[metricLabel]),
      name: joinedName,
      itemStyle: {
        color: colorFn(joinedName),
        opacity: isFiltered ? _constants.OpacityEnum.Transparent : _constants.OpacityEnum.NonTransparent
      },
      lineStyle: {
        opacity: isFiltered ? _constants.OpacityEnum.SemiTransparent : _constants.OpacityEnum.NonTransparent
      },
      label: {
        show: showLabels,
        position: labelPosition,
        formatter
      }
    });
  });
  const selectedValues = (filterState.selectedValues || []).reduce((acc, selectedValue) => {
    const index = transformedData.findIndex(({
      name
    }) => name === selectedValue);
    return { ...acc,
      [index]: selectedValue
    };
  }, {});
  const indicator = metricsLabel.map(metricLabel => {
    var _columnConfig$metricL;

    const maxValueInControl = columnConfig == null ? void 0 : (_columnConfig$metricL = columnConfig[metricLabel]) == null ? void 0 : _columnConfig$metricL.radarMetricMaxValue; // Ensure that 0 is at the center of the polar coordinates

    const metricValueAsMax = metricLabelAndMaxValueMap.get(metricLabel) === 0 ? Number.MAX_SAFE_INTEGER : metricLabelAndMaxValueMap.get(metricLabel);
    const max = maxValueInControl === null ? metricValueAsMax : maxValueInControl;
    return {
      name: metricLabel,
      max
    };
  });
  const series = [{
    type: 'radar',
    animation: false,
    emphasis: {
      label: {
        show: true,
        fontWeight: 'bold',
        backgroundColor: 'white'
      }
    },
    data: transformedData
  }];
  const echartOptions = {
    grid: { ..._defaults.defaultGrid
    },
    tooltip: { ..._defaults.defaultTooltip,
      trigger: 'item'
    },
    legend: { ...(0, _series.getLegendProps)(legendType, legendOrientation, showLegend),
      data: Array.from(columnsLabelMap.keys())
    },
    series,
    radar: {
      shape: isCircle ? 'circle' : 'polygon',
      indicator
    }
  };
  return {
    formData,
    width,
    height,
    echartOptions,
    setDataMask,
    labelMap: Object.fromEntries(columnsLabelMap),
    groupby,
    selectedValues
  };
}
