// @vx/threshold/build/Threshold.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Threshold;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _shape = require('@vx/shape');

var _clipPath = require('@vx/clip-path');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Threshold(_ref) {
  var className = _ref.className,
      curve = _ref.curve,
      xScale = _ref.xScale,
      yScale = _ref.yScale,
      clipAboveTo = _ref.clipAboveTo,
      clipBelowTo = _ref.clipBelowTo,
      data = _ref.data,
      x = _ref.x,
      y0 = _ref.y0,
      y1 = _ref.y1,
      aboveAreaProps = _ref.aboveAreaProps,
      belowAreaProps = _ref.belowAreaProps;

  return _react2.default.createElement(
    'g',
    { className: (0, _classnames2.default)('vx-threshold', className) },
    _react2.default.createElement(
      _shape.Area,
      { curve: curve, data: data, x: x, y1: y1, xScale: xScale, yScale: yScale },
      function (_ref2) {
        var path = _ref2.path;

        return _react2.default.createElement(
          'g',
          null,
          _react2.default.createElement(
            _clipPath.ClipPath,
            { id: 'threshold-clip-below' },
            _react2.default.createElement('path', { d: path.y0(clipBelowTo)(data) })
          ),
          _react2.default.createElement(
            _clipPath.ClipPath,
            { id: 'threshold-clip-above' },
            _react2.default.createElement('path', { d: path.y0(clipAboveTo)(data) })
          )
        );
      }
    ),
    _react2.default.createElement(_shape.Area, _extends({
      curve: curve,
      data: data,
      x: x,
      y0: y0,
      y1: y1,
      xScale: xScale,
      yScale: yScale,
      strokeWidth: 0,
      clipPath: 'url(#threshold-clip-below)'
    }, belowAreaProps)),
    _react2.default.createElement(_shape.Area, _extends({
      curve: curve,
      data: data,
      x: x,
      y0: y0,
      y1: y1,
      xScale: xScale,
      yScale: yScale,
      strokeWidth: 0,
      clipPath: 'url(#threshold-clip-above)'
    }, aboveAreaProps))
  );
}
