// @superset-ui/legacy-plugin-chart-rose/lib/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const config = {
  controlPanelSections: [_chartControls.sections.legacyTimeseriesTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [['metrics'], ['adhoc_filters'], ['groupby'], ['limit', 'timeseries_limit_metric'], [{
      name: 'order_desc',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Sort Descending'),
        default: true,
        description: (0, _core.t)('Whether to sort descending or ascending')
      }
    }, {
      name: 'contribution',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Contribution'),
        default: false,
        description: (0, _core.t)('Compute the contribution to the total')
      }
    }], ['row_limit', null]]
  }, {
    label: (0, _core.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_scheme'], [{
      name: 'number_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Number format'),
        renderTrigger: true,
        default: 'SMART_NUMBER',
        choices: _chartControls.D3_FORMAT_OPTIONS,
        description: _chartControls.D3_FORMAT_DOCS
      }
    }, {
      name: 'date_time_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Date Time Format'),
        renderTrigger: true,
        default: 'smart_date',
        choices: _chartControls.D3_TIME_FORMAT_OPTIONS,
        description: _chartControls.D3_FORMAT_DOCS
      }
    }], [{
      name: 'rich_tooltip',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Rich Tooltip'),
        renderTrigger: true,
        default: true,
        description: (0, _core.t)('The rich tooltip shows a list of all series for that point in time')
      }
    }, {
      name: 'rose_area_proportion',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Use Area Proportions'),
        description: (0, _core.t)('Check if the Rose Chart should use segment area instead of ' + 'segment radius for proportioning'),
        default: false,
        renderTrigger: true
      }
    }]]
  }, {
    label: (0, _core.t)('Advanced Analytics'),
    tabOverride: 'data',
    description: (0, _core.t)('This section contains options ' + 'that allow for advanced analytical post processing ' + 'of query results'),
    controlSetRows: [// eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Rolling Window'))], [{
      name: 'rolling_type',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Rolling Function'),
        default: 'None',
        choices: (0, _chartControls.formatSelectOptions)(['None', 'mean', 'sum', 'std', 'cumsum']),
        description: (0, _core.t)('Defines a rolling window function to apply, works along ' + 'with the [Periods] text box')
      }
    }], [{
      name: 'rolling_periods',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Periods'),
        isInt: true,
        description: (0, _core.t)('Defines the size of the rolling window function, ' + 'relative to the time granularity selected')
      }
    }, {
      name: 'min_periods',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Min Periods'),
        isInt: true,
        description: (0, _core.t)('The minimum number of rolling periods required to show ' + 'a value. For instance if you do a cumulative sum on 7 days ' + 'you may want your "Min Period" to be 7, so that all data points ' + 'shown are the total of 7 periods. This will hide the "ramp up" ' + 'taking place over the first 7 periods')
      }
    }], // eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Time Comparison'))], [{
      name: 'time_compare',
      config: {
        type: 'SelectControl',
        multi: true,
        freeForm: true,
        label: (0, _core.t)('Time Shift'),
        choices: (0, _chartControls.formatSelectOptions)(['1 day', '1 week', '28 days', '30 days', '52 weeks', '1 year', '104 weeks', '2 years']),
        description: (0, _core.t)('Overlay one or more timeseries from a ' + 'relative time period. Expects relative time deltas ' + 'in natural language (example:  24 hours, 7 days, ' + '52 weeks, 365 days). Free text is supported.')
      }
    }, {
      name: 'comparison_type',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Calculation type'),
        default: 'values',
        choices: [['values', 'Actual Values'], ['absolute', 'Difference'], ['percentage', 'Percentage change'], ['ratio', 'Ratio']],
        description: (0, _core.t)('How to display time shifts: as individual lines; as the ' + 'difference between the main time series and each time shift; ' + 'as the percentage change; or as the ratio between series and time shifts.')
      }
    }], // eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Python Functions'))], // eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h2", {
      className: "section-header"
    }, "pandas.resample")], [{
      name: 'resample_rule',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Rule'),
        default: null,
        choices: (0, _chartControls.formatSelectOptions)(['1T', '1H', '1D', '7D', '1M', '1AS']),
        description: (0, _core.t)('Pandas resample rule')
      }
    }, {
      name: 'resample_method',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Method'),
        default: null,
        choices: (0, _chartControls.formatSelectOptions)(['asfreq', 'bfill', 'ffill', 'median', 'mean', 'sum']),
        description: (0, _core.t)('Pandas resample method')
      }
    }]]
  }]
};
var _default = config;
exports.default = _default;
