// @applitools/logger/dist/printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makePrinter = void 0;
const log_level_1 = require("./log-level");
function makePrinter({ handler, format, level, ...defaults }) {
    return { debug, log, warn, error, fatal, verbose: log };
    function debug(...messages) {
        if (level < log_level_1.LogLevel.debug)
            return;
        const options = { ...defaults, level: 'debug' };
        handler.log(format(messages, options));
    }
    function log(...messages) {
        if (level < log_level_1.LogLevel.info)
            return;
        const options = { ...defaults, level: 'info' };
        handler.log(format(messages, options));
    }
    function warn(...messages) {
        if (level < log_level_1.LogLevel.warn)
            return;
        const options = { ...defaults, level: 'warn' };
        if (handler.warn)
            handler.warn(format(messages, options));
        else
            handler.log(format(messages, options));
    }
    function error(...messages) {
        if (level < log_level_1.LogLevel.error)
            return;
        const options = { ...defaults, level: 'error' };
        if (handler.error)
            handler.error(format(messages, options));
        else
            handler.log(format(messages, options));
    }
    function fatal(...messages) {
        if (level < log_level_1.LogLevel.fatal)
            return;
        const options = { ...defaults, level: 'fatal' };
        if (handler.fatal)
            handler.fatal(format(messages, options));
        else if (handler.error)
            handler.error(format(messages, options));
        else
            handler.log(format(messages, options));
    }
}
exports.makePrinter = makePrinter;
