// @encodable/color/lib/scale/ColorNamespace.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _registry = require("@encodable/registry");

var _scheme = require("../scheme");

var _stringifyAndTrim = _interopRequireDefault(require("../utils/stringifyAndTrim"));

var _ScaleCategoricalColor = _interopRequireDefault(require("./ScaleCategoricalColor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ColorNamespace {
  constructor(nameOrState) {
    _defineProperty(this, "state", void 0);

    _defineProperty(this, "scales", void 0);

    this.state = typeof nameOrState === 'string' ? {
      name: nameOrState,
      manualColors: {},
      scales: {}
    } : nameOrState;
    this.scales = new _registry.SyncRegistry();
  }

  get name() {
    return this.state.name;
  }
  /**
   * Check if the specified value is subjected to manual color assignment
   * @param value
   */


  hasManualColor(value) {
    return typeof this.state.manualColors[(0, _stringifyAndTrim.default)(value)] !== 'undefined';
  }
  /**
   * Enforce specific color for given value
   * This will apply across all color scales
   * in this namespace.
   * @param {*} value value
   * @param {*} manualColor color
   */


  setColor(value, manualColor) {
    this.state.manualColors[(0, _stringifyAndTrim.default)(value)] = manualColor;
    return this;
  }
  /**
   * Removed enforced specific color for given value
   * This will apply across all color scales
   * in this namespace.
   * @param {*} value value
   * @param {*} manualColor color
   */


  unsetColor(value) {
    delete this.state.manualColors[(0, _stringifyAndTrim.default)(value)];
    return this;
  }
  /**
   * Clear all manually assigned colors
   */


  clearManualColors() {
    this.state.manualColors = {};
    return this;
  }

  hasScale(scheme) {
    return typeof this.state.scales[scheme] !== 'undefined';
  }

  getScale(scheme) {
    var _ref;

    const schemeName = (_ref = scheme != null ? scheme : (0, _scheme.getColorSchemeRegistry)().categorical.getDefaultKey()) != null ? _ref : 'undefined';

    if (typeof this.state.scales[schemeName] === 'undefined') {
      var _getCategoricalScheme, _getCategoricalScheme2;

      // create scale
      const scale = new _ScaleCategoricalColor.default((_getCategoricalScheme = (_getCategoricalScheme2 = (0, _scheme.getCategoricalScheme)(schemeName)) == null ? void 0 : _getCategoricalScheme2.colors.slice()) != null ? _getCategoricalScheme : [], this.state.manualColors); // add state to lookup

      this.state.scales[schemeName] = scale.state;
      this.scales.registerValue(schemeName, scale);
      return scale;
    }

    if (this.scales.has(schemeName)) {
      return this.scales.get(schemeName);
    } // create scale


    const scale = new _ScaleCategoricalColor.default(this.state.scales[schemeName], this.state.manualColors);
    this.scales.registerValue(schemeName, scale);
    return scale;
  }

}

exports.default = ColorNamespace;
