// @encodable/color/lib/scale/index.js
"use strict";

exports.__esModule = true;
exports.getColorNamespace = getColorNamespace;
exports.getScale = getScale;
exports.getColor = getColor;
exports.getColorNamespaceRegistry = void 0;

var _registry = require("@encodable/registry");

var _ColorNamespaceRegistry = _interopRequireDefault(require("./ColorNamespaceRegistry"));

exports.ColorNamespaceRegistry = _ColorNamespaceRegistry.default;

var _stringifyAndTrim = _interopRequireDefault(require("../utils/stringifyAndTrim"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getColorNamespaceRegistry = (0, _registry.makeSingleton)(() => new _ColorNamespaceRegistry.default({
  globalId: '@encodable/color:ColorNamespaceRegistry'
}));
exports.getColorNamespaceRegistry = getColorNamespaceRegistry;

function getColorNamespace(namespace) {
  return getColorNamespaceRegistry().get(namespace);
}

function getScale(scheme, namespace) {
  return getColorNamespaceRegistry().get(namespace).getScale(scheme);
}

function getColor(value, scheme, namespace) {
  return getColorNamespaceRegistry().get(namespace).getScale(scheme)((0, _stringifyAndTrim.default)(value));
}
