// @superset-ui/core/lib/chart/models/ChartPlugin.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = require("../..");

var _ChartMetadataRegistrySingleton = _interopRequireDefault(require("../registries/ChartMetadataRegistrySingleton"));

var _ChartBuildQueryRegistrySingleton = _interopRequireDefault(require("../registries/ChartBuildQueryRegistrySingleton"));

var _ChartComponentRegistrySingleton = _interopRequireDefault(require("../registries/ChartComponentRegistrySingleton"));

var _ChartControlPanelRegistrySingleton = _interopRequireDefault(require("../registries/ChartControlPanelRegistrySingleton"));

var _ChartTransformPropsRegistrySingleton = _interopRequireDefault(require("../registries/ChartTransformPropsRegistrySingleton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
function IDENTITY(x) {
  return x;
}

const EMPTY = {};

/**
 * Loaders of the form `() => import('foo')` may return esmodules
 * which require the value to be extracted as `module.default`
 * */
function sanitizeLoader(loader) {
  return () => {
    const loaded = loader();
    return loaded instanceof Promise ? loaded.then(module => 'default' in module && module.default || module) : loaded;
  };
}

class ChartPlugin extends _.Plugin {
  constructor(config) {
    super();
    this.controlPanel = void 0;
    this.metadata = void 0;
    this.loadBuildQuery = void 0;
    this.loadTransformProps = void 0;
    this.loadChart = void 0;
    const {
      metadata,
      buildQuery,
      loadBuildQuery,
      transformProps = IDENTITY,
      loadTransformProps,
      Chart,
      loadChart,
      controlPanel = EMPTY
    } = config;
    this.controlPanel = controlPanel;
    this.metadata = metadata;
    this.loadBuildQuery = loadBuildQuery && sanitizeLoader(loadBuildQuery) || buildQuery && sanitizeLoader(() => buildQuery) || undefined;
    this.loadTransformProps = sanitizeLoader(loadTransformProps ?? (() => transformProps));

    if (loadChart) {
      this.loadChart = sanitizeLoader(loadChart);
    } else if (Chart) {
      this.loadChart = () => Chart;
    } else {
      throw new Error('Chart or loadChart is required');
    }
  }

  register() {
    const key = this.config.key || (0, _.isRequired)('config.key');
    (0, _ChartMetadataRegistrySingleton.default)().registerValue(key, this.metadata);
    (0, _ChartComponentRegistrySingleton.default)().registerLoader(key, this.loadChart);
    (0, _ChartControlPanelRegistrySingleton.default)().registerValue(key, this.controlPanel);
    (0, _ChartTransformPropsRegistrySingleton.default)().registerLoader(key, this.loadTransformProps);

    if (this.loadBuildQuery) {
      (0, _ChartBuildQueryRegistrySingleton.default)().registerLoader(key, this.loadBuildQuery);
    }

    return this;
  }

  unregister() {
    const key = this.config.key || (0, _.isRequired)('config.key');
    (0, _ChartMetadataRegistrySingleton.default)().remove(key);
    (0, _ChartComponentRegistrySingleton.default)().remove(key);
    (0, _ChartControlPanelRegistrySingleton.default)().remove(key);
    (0, _ChartTransformPropsRegistrySingleton.default)().remove(key);
    (0, _ChartBuildQueryRegistrySingleton.default)().remove(key);
    return this;
  }

  configure(config, replace) {
    super.configure(config, replace);
    return this;
  }

}

exports.default = ChartPlugin;
