// @superset-ui/plugin-chart-echarts/lib/constants.js
"use strict";

exports.__esModule = true;
exports.TIMESERIES_CONSTANTS = exports.OpacityEnum = exports.NULL_STRING = exports.LABEL_POSITION = void 0;

var _types = require("./types");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// eslint-disable-next-line import/prefer-default-export
const NULL_STRING = '<NULL>';
exports.NULL_STRING = NULL_STRING;
const TIMESERIES_CONSTANTS = {
  gridOffsetRight: 40,
  gridOffsetLeft: 20,
  gridOffsetTop: 20,
  gridOffsetBottom: 20,
  gridOffsetBottomZoomable: 80,
  legendRightTopOffset: 30,
  legendTopRightOffset: 55,
  zoomBottom: 30,
  toolboxTop: 0,
  toolboxRight: 5,
  dataZoomStart: 0,
  dataZoomEnd: 100,
  yAxisLabelTopOffset: 20
};
exports.TIMESERIES_CONSTANTS = TIMESERIES_CONSTANTS;
const LABEL_POSITION = [[_types.LabelPositionEnum.Top, 'Top'], [_types.LabelPositionEnum.Left, 'Left'], [_types.LabelPositionEnum.Right, 'Right'], [_types.LabelPositionEnum.Bottom, 'Bottom'], [_types.LabelPositionEnum.Inside, 'Inside'], [_types.LabelPositionEnum.InsideBottomLeft, 'Inside left'], [_types.LabelPositionEnum.InsideBottomRight, 'Inside right'], [_types.LabelPositionEnum.InsideTop, 'Inside top'], [_types.LabelPositionEnum.InsideBottom, 'Inside bottom'], [_types.LabelPositionEnum.InsideTopLeft, 'Inside top left'], [_types.LabelPositionEnum.InsideBottomLeft, 'Inside bottom left'], [_types.LabelPositionEnum.InsideTopRight, 'Inside top right'], [_types.LabelPositionEnum.InsideBottomRight, 'Inside bottom right']];
exports.LABEL_POSITION = LABEL_POSITION;
let OpacityEnum;
exports.OpacityEnum = OpacityEnum;

(function (OpacityEnum) {
  OpacityEnum[OpacityEnum["Transparent"] = 0] = "Transparent";
  OpacityEnum[OpacityEnum["SemiTransparent"] = 0.3] = "SemiTransparent";
  OpacityEnum[OpacityEnum["NonTransparent"] = 1] = "NonTransparent";
})(OpacityEnum || (exports.OpacityEnum = OpacityEnum = {}));
