// @superset-ui/plugin-chart-echarts/lib/controls.js
"use strict";

exports.__esModule = true;
exports.showValueSection = exports.richTooltipSection = exports.legendSection = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

var _types = require("./types");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const {
  legendMargin,
  legendOrientation,
  legendType,
  showLegend
} = _types.DEFAULT_LEGEND_FORM_DATA;
const showLegendControl = {
  name: 'show_legend',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Show legend'),
    renderTrigger: true,
    default: showLegend,
    description: (0, _core.t)('Whether to display a legend for the chart')
  }
};
const legendMarginControl = {
  name: 'legendMargin',
  config: {
    type: 'TextControl',
    label: (0, _core.t)('Margin'),
    renderTrigger: true,
    isInt: true,
    default: legendMargin,
    description: (0, _core.t)('Additional padding for legend.'),
    visibility: ({
      controls
    }) => {
      var _controls$show_legend;

      return Boolean(controls == null ? void 0 : (_controls$show_legend = controls.show_legend) == null ? void 0 : _controls$show_legend.value);
    }
  }
};
const legendTypeControl = {
  name: 'legendType',
  config: {
    type: 'SelectControl',
    freeForm: false,
    label: 'Type',
    choices: [['scroll', 'Scroll'], ['plain', 'Plain']],
    default: legendType,
    renderTrigger: true,
    description: (0, _core.t)('Legend type'),
    visibility: ({
      controls
    }) => {
      var _controls$show_legend2;

      return Boolean(controls == null ? void 0 : (_controls$show_legend2 = controls.show_legend) == null ? void 0 : _controls$show_legend2.value);
    }
  }
};
const legendOrientationControl = {
  name: 'legendOrientation',
  config: {
    type: 'SelectControl',
    freeForm: false,
    label: 'Orientation',
    choices: [['top', 'Top'], ['bottom', 'Bottom'], ['left', 'Left'], ['right', 'Right']],
    default: legendOrientation,
    renderTrigger: true,
    description: (0, _core.t)('Legend type'),
    visibility: ({
      controls
    }) => {
      var _controls$show_legend3;

      return Boolean(controls == null ? void 0 : (_controls$show_legend3 = controls.show_legend) == null ? void 0 : _controls$show_legend3.value);
    }
  }
};
const legendSection = [[(0, _react2.jsx)("h1", {
  className: "section-header"
}, (0, _core.t)('Legend'))], [showLegendControl], [legendTypeControl], [legendOrientationControl], [legendMarginControl]];
exports.legendSection = legendSection;
const showValueControl = {
  name: 'show_value',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Show Value'),
    default: false,
    renderTrigger: true,
    description: (0, _core.t)('Show series values on the chart')
  }
};
const stackControl = {
  name: 'stack',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Stack series'),
    renderTrigger: true,
    default: false,
    description: (0, _core.t)('Stack series on top of each other')
  }
};
const onlyTotalControl = {
  name: 'only_total',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Only Total'),
    default: true,
    renderTrigger: true,
    description: (0, _core.t)('Only show the total value on the stacked chart, and not show on the selected category'),
    visibility: ({
      controls
    }) => {
      var _controls$show_value, _controls$stack;

      return Boolean(controls == null ? void 0 : (_controls$show_value = controls.show_value) == null ? void 0 : _controls$show_value.value) && Boolean(controls == null ? void 0 : (_controls$stack = controls.stack) == null ? void 0 : _controls$stack.value);
    }
  }
};
const showValueSection = [[showValueControl], [stackControl], [onlyTotalControl]];
exports.showValueSection = showValueSection;
const richTooltipControl = {
  name: 'rich_tooltip',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Rich tooltip'),
    renderTrigger: true,
    default: true,
    description: (0, _core.t)('Shows a list of all series available at that point in time')
  }
};
const tooltipTimeFormatControl = {
  name: 'tooltipTimeFormat',
  config: { ..._chartControls.sharedControls.x_axis_time_format,
    label: (0, _core.t)('Tooltip time format'),
    default: 'smart_date',
    clearable: false
  }
};
const tooltipSortByMetricControl = {
  name: 'tooltipSortByMetric',
  config: {
    type: 'CheckboxControl',
    label: (0, _core.t)('Tooltip sort by metric'),
    renderTrigger: true,
    default: false,
    description: (0, _core.t)('Whether to sort tooltip by the selected metric in descending order.'),
    visibility: ({
      controls
    }) => {
      var _controls$rich_toolti;

      return Boolean(controls == null ? void 0 : (_controls$rich_toolti = controls.rich_tooltip) == null ? void 0 : _controls$rich_toolti.value);
    }
  }
};
const richTooltipSection = [[(0, _react2.jsx)("h1", {
  className: "section-header"
}, (0, _core.t)('Tooltip'))], [richTooltipControl], [tooltipSortByMetricControl], [tooltipTimeFormatControl]];
exports.richTooltipSection = richTooltipSection;
