// @superset-ui/plugin-chart-echarts/lib/types.js
"use strict";

exports.__esModule = true;
exports.LegendType = exports.LegendOrientation = exports.LabelPositionEnum = exports.ForecastSeriesEnum = exports.DEFAULT_TITLE_FORM_DATA = exports.DEFAULT_LEGEND_FORM_DATA = void 0;

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
let ForecastSeriesEnum;
exports.ForecastSeriesEnum = ForecastSeriesEnum;

(function (ForecastSeriesEnum) {
  ForecastSeriesEnum["Observation"] = "";
  ForecastSeriesEnum["ForecastTrend"] = "__yhat";
  ForecastSeriesEnum["ForecastUpper"] = "__yhat_upper";
  ForecastSeriesEnum["ForecastLower"] = "__yhat_lower";
})(ForecastSeriesEnum || (exports.ForecastSeriesEnum = ForecastSeriesEnum = {}));

let LegendOrientation;
exports.LegendOrientation = LegendOrientation;

(function (LegendOrientation) {
  LegendOrientation["Top"] = "top";
  LegendOrientation["Bottom"] = "bottom";
  LegendOrientation["Left"] = "left";
  LegendOrientation["Right"] = "right";
})(LegendOrientation || (exports.LegendOrientation = LegendOrientation = {}));

let LegendType;
exports.LegendType = LegendType;

(function (LegendType) {
  LegendType["Scroll"] = "scroll";
  LegendType["Plain"] = "plain";
})(LegendType || (exports.LegendType = LegendType = {}));

const DEFAULT_LEGEND_FORM_DATA = {
  legendMargin: null,
  legendOrientation: LegendOrientation.Top,
  legendType: LegendType.Scroll,
  showLegend: false
};
exports.DEFAULT_LEGEND_FORM_DATA = DEFAULT_LEGEND_FORM_DATA;
let LabelPositionEnum;
exports.LabelPositionEnum = LabelPositionEnum;

(function (LabelPositionEnum) {
  LabelPositionEnum["Top"] = "top";
  LabelPositionEnum["Left"] = "left";
  LabelPositionEnum["Right"] = "right";
  LabelPositionEnum["Bottom"] = "bottom";
  LabelPositionEnum["Inside"] = "inside";
  LabelPositionEnum["InsideLeft"] = "insideLeft";
  LabelPositionEnum["InsideRight"] = "insideRight";
  LabelPositionEnum["InsideTop"] = "insideTop";
  LabelPositionEnum["InsideBottom"] = "insideBottom";
  LabelPositionEnum["InsideTopLeft"] = "insideTopLeft";
  LabelPositionEnum["InsideBottomLeft"] = "insideBottomLeft";
  LabelPositionEnum["InsideTopRight"] = "insideTopRight";
  LabelPositionEnum["InsideBottomRight"] = "insideBottomRight";
})(LabelPositionEnum || (exports.LabelPositionEnum = LabelPositionEnum = {}));

const DEFAULT_TITLE_FORM_DATA = {
  xAxisTitle: '',
  xAxisTitleMargin: 0,
  yAxisTitle: '',
  yAxisTitleMargin: 0,
  yAxisTitlePosition: 'Top'
};
exports.DEFAULT_TITLE_FORM_DATA = DEFAULT_TITLE_FORM_DATA;
