// @superset-ui/core/lib/query/constants.js
"use strict";

exports.__esModule = true;
exports.EXTRA_FORM_DATA_OVERRIDE_REGULAR_MAPPINGS = exports.EXTRA_FORM_DATA_OVERRIDE_REGULAR_KEYS = exports.EXTRA_FORM_DATA_OVERRIDE_KEYS = exports.EXTRA_FORM_DATA_OVERRIDE_EXTRA_KEYS = exports.EXTRA_FORM_DATA_APPEND_KEYS = exports.DTTM_ALIAS = void 0;

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const DTTM_ALIAS = '__timestamp';
exports.DTTM_ALIAS = DTTM_ALIAS;
const EXTRA_FORM_DATA_OVERRIDE_EXTRA_KEYS = ['druid_time_origin', 'relative_start', 'relative_end', 'time_grain_sqla', 'time_range_endpoints'];
exports.EXTRA_FORM_DATA_OVERRIDE_EXTRA_KEYS = EXTRA_FORM_DATA_OVERRIDE_EXTRA_KEYS;
const EXTRA_FORM_DATA_APPEND_KEYS = ['adhoc_filters', 'filters', 'interactive_groupby', 'interactive_highlight', 'interactive_drilldown', 'custom_form_data'];
exports.EXTRA_FORM_DATA_APPEND_KEYS = EXTRA_FORM_DATA_APPEND_KEYS;
const EXTRA_FORM_DATA_OVERRIDE_REGULAR_MAPPINGS = {
  granularity: 'granularity',
  granularity_sqla: 'granularity',
  time_column: 'time_column',
  time_grain: 'time_grain',
  time_range: 'time_range'
};
exports.EXTRA_FORM_DATA_OVERRIDE_REGULAR_MAPPINGS = EXTRA_FORM_DATA_OVERRIDE_REGULAR_MAPPINGS;
const EXTRA_FORM_DATA_OVERRIDE_REGULAR_KEYS = Object.keys(EXTRA_FORM_DATA_OVERRIDE_REGULAR_MAPPINGS);
exports.EXTRA_FORM_DATA_OVERRIDE_REGULAR_KEYS = EXTRA_FORM_DATA_OVERRIDE_REGULAR_KEYS;
const EXTRA_FORM_DATA_OVERRIDE_KEYS = [...EXTRA_FORM_DATA_OVERRIDE_REGULAR_KEYS, ...EXTRA_FORM_DATA_OVERRIDE_EXTRA_KEYS];
exports.EXTRA_FORM_DATA_OVERRIDE_KEYS = EXTRA_FORM_DATA_OVERRIDE_KEYS;
