// @encodable/color/lib/scheme/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  getColorSchemeRegistry: true,
  getColorScheme: true,
  getCategoricalScheme: true,
  getSequentialScheme: true,
  getDivergingScheme: true,
  wrapColorScheme: true,
  ColorSchemeRegistry: true
};
exports.getColorScheme = getColorScheme;
exports.getCategoricalScheme = getCategoricalScheme;
exports.getSequentialScheme = getSequentialScheme;
exports.getDivergingScheme = getDivergingScheme;
exports.wrapColorScheme = exports.getColorSchemeRegistry = void 0;

var _registry = require("@encodable/registry");

var _ColorSchemeRegistry = _interopRequireDefault(require("./ColorSchemeRegistry"));

exports.ColorSchemeRegistry = _ColorSchemeRegistry.default;

var _d3Schemes = require("./presets/d3Schemes");

Object.keys(_d3Schemes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _d3Schemes[key];
});

var _wrapColorScheme = _interopRequireDefault(require("./wrappers/wrapColorScheme"));

exports.wrapColorScheme = _wrapColorScheme.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getColorSchemeRegistry = (0, _registry.makeSingleton)(() => new _ColorSchemeRegistry.default({
  globalId: '@encodable/color:ColorSchemeRegistry'
}).register(_d3Schemes.d3Schemes));
exports.getColorSchemeRegistry = getColorSchemeRegistry;

function getColorScheme(schemeId) {
  return getColorSchemeRegistry().get(schemeId);
}

function getCategoricalScheme(schemeId) {
  return getColorSchemeRegistry().categorical.get(schemeId);
}

function getSequentialScheme(schemeId) {
  return getColorSchemeRegistry().sequential.get(schemeId);
}

function getDivergingScheme(schemeId) {
  return getColorSchemeRegistry().diverging.get(schemeId);
}
