// @applitools/core/dist/ufg/check-and-close.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCheckAndClose = void 0;
const driver_1 = require("@applitools/driver");
const take_snapshots_1 = require("./utils/take-snapshots");
const wait_for_lazy_load_1 = require("../utils/wait-for-lazy-load");
const to_base_check_settings_1 = require("../utils/to-base-check-settings");
const generate_safe_selectors_1 = require("./utils/generate-safe-selectors");
const abort_error_1 = require("../errors/abort-error");
const utils = __importStar(require("@applitools/utils"));
const chalk_1 = __importDefault(require("chalk"));
function makeCheckAndClose({ spec, eyes, client, signal, target: defaultTarget, logger: defaultLogger, }) {
    return async function checkAndClose({ target = defaultTarget, settings = {}, logger = defaultLogger, snapshots, }) {
        var _a, _b;
        logger.log('Command "checkAndClose" is called with settings', settings);
        if (signal.aborted) {
            logger.warn('Command "checkAndClose" was called after test was already aborted');
            throw new abort_error_1.AbortError('Command "checkAndClose" was called after test was already aborted');
        }
        const { elementReferencesToCalculate, elementReferenceToTarget, getBaseCheckSettings } = (0, to_base_check_settings_1.toBaseCheckSettings)({ settings });
        let snapshotUrl, snapshotTitle, userAgent, regionToTarget, selectorsToCalculate;
        if ((0, driver_1.isDriver)(target, spec)) {
            // TODO driver custom config
            const driver = await (0, driver_1.makeDriver)({ spec, driver: target, logger });
            if (driver.isWeb && (!settings.renderers || settings.renderers.length === 0)) {
                const viewportSize = await driver.getViewportSize();
                settings.renderers = [{ name: 'chrome', ...viewportSize }];
            }
            let cleanupGeneratedSelectors;
            if (driver.isWeb) {
                userAgent = driver.userAgent;
                const generated = await (0, generate_safe_selectors_1.generateSafeSelectors)({
                    context: driver.currentContext,
                    elementReferences: [...(elementReferenceToTarget ? [elementReferenceToTarget] : []), ...elementReferencesToCalculate],
                });
                cleanupGeneratedSelectors = generated.cleanupGeneratedSelectors;
                if (elementReferenceToTarget) {
                    regionToTarget = (_a = generated.selectors[0]) === null || _a === void 0 ? void 0 : _a.safeSelector;
                    if (!regionToTarget)
                        throw new Error('Target element not found');
                    selectorsToCalculate = generated.selectors.slice(1);
                }
                else {
                    selectorsToCalculate = generated.selectors;
                }
            }
            const currentContext = driver.currentContext;
            if (!snapshots) {
                snapshots = await (0, take_snapshots_1.takeSnapshots)({
                    driver,
                    settings: {
                        ...eyes.test.server,
                        waitBeforeCapture: settings.waitBeforeCapture,
                        disableBrowserFetching: settings.disableBrowserFetching,
                        layoutBreakpoints: settings.layoutBreakpoints,
                        renderers: settings.renderers,
                        skipResources: client.getCachedResourceUrls(),
                    },
                    hooks: {
                        async beforeSnapshots() {
                            if (driver.isWeb && settings.lazyLoad) {
                                await (0, wait_for_lazy_load_1.waitForLazyLoad)({
                                    context: driver.currentContext,
                                    settings: settings.lazyLoad !== true ? settings.lazyLoad : {},
                                    logger,
                                });
                            }
                        },
                    },
                    provides: {
                        getChromeEmulationDevices: client.getChromeEmulationDevices,
                        getIOSDevices: client.getIOSDevices,
                    },
                    logger,
                });
            }
            await currentContext.focus();
            snapshotUrl = await driver.getUrl();
            snapshotTitle = await driver.getTitle();
            await (cleanupGeneratedSelectors === null || cleanupGeneratedSelectors === void 0 ? void 0 : cleanupGeneratedSelectors());
        }
        else {
            snapshots = !utils.types.isArray(target) ? Array(settings.renderers.length).fill(target) : target;
            snapshotUrl = utils.types.has(snapshots[0], 'url') ? snapshots[0].url : undefined;
        }
        regionToTarget !== null && regionToTarget !== void 0 ? regionToTarget : (regionToTarget = (_b = elementReferenceToTarget) !== null && _b !== void 0 ? _b : settings.region);
        selectorsToCalculate !== null && selectorsToCalculate !== void 0 ? selectorsToCalculate : (selectorsToCalculate = elementReferencesToCalculate.map(selector => ({
            originalSelector: selector,
            safeSelector: selector,
        })));
        const promises = settings.renderers.map(async (renderer, index) => {
            if (utils.types.has(renderer, 'name') && renderer.name === 'edge') {
                const message = chalk_1.default.yellow(`The 'edge' option that is being used in your browsers' configuration will soon be deprecated. Please change it to either 'edgelegacy' for the legacy version or to 'edgechromium' for the new Chromium-based version. Please note, when using the built-in BrowserType enum, then the values are BrowserType.EDGE_LEGACY and BrowserType.EDGE_CHROMIUM, respectively.`);
                logger.console.log(message);
            }
            try {
                if (signal.aborted) {
                    logger.warn('Command "check" was aborted before rendering');
                    throw new abort_error_1.AbortError('Command "check" was aborted before rendering');
                }
                const { cookies, ...snapshot } = snapshots[index];
                const snapshotType = utils.types.has(snapshot, 'cdt') ? 'web' : 'native';
                const renderTargetPromise = client.createRenderTarget({
                    snapshot,
                    settings: {
                        renderer,
                        referer: snapshotUrl,
                        cookies,
                        proxy: eyes.test.server.proxy,
                        autProxy: settings.autProxy,
                        userAgent,
                    },
                });
                const [baseEyes] = await eyes.getBaseEyes({ settings: { renderer, type: snapshotType }, logger });
                try {
                    if (signal.aborted) {
                        logger.warn('Command "check" was aborted before rendering');
                        throw new abort_error_1.AbortError('Command "check" was aborted before rendering');
                    }
                    else if (baseEyes.aborted) {
                        logger.warn(`Renderer with id ${baseEyes.test.rendererId} was aborted during one of the previous steps`);
                        throw new abort_error_1.AbortError(`Renderer with id "${baseEyes.test.rendererId}" was aborted during one of the previous steps`);
                    }
                    const renderTarget = await renderTargetPromise;
                    if (signal.aborted) {
                        logger.warn('Command "check" was aborted before rendering');
                        throw new abort_error_1.AbortError('Command "check" was aborted before rendering');
                    }
                    else if (baseEyes.aborted) {
                        logger.warn(`Renderer with id ${baseEyes.test.rendererId} was aborted during one of the previous steps`);
                        throw new abort_error_1.AbortError(`Renderer with id "${baseEyes.test.rendererId}" was aborted during one of the previous steps`);
                    }
                    const { renderId, selectorRegions, ...baseTarget } = await client.render({
                        target: renderTarget,
                        settings: {
                            ...settings,
                            rendererId: baseEyes.test.rendererId,
                            region: regionToTarget,
                            type: utils.types.has(snapshot, 'cdt') ? 'web' : 'native',
                            renderer,
                            selectorsToCalculate: selectorsToCalculate.flatMap(({ safeSelector }) => safeSelector !== null && safeSelector !== void 0 ? safeSelector : []),
                            includeFullPageSize: Boolean(settings.pageId),
                        },
                        signal,
                    });
                    let offset = 0;
                    const baseSettings = getBaseCheckSettings({
                        calculatedRegions: selectorsToCalculate.map(({ originalSelector, safeSelector }) => ({
                            selector: originalSelector,
                            regions: safeSelector ? selectorRegions[offset++] : [],
                        })),
                    });
                    baseSettings.renderId = renderId;
                    baseTarget.source = snapshotUrl;
                    baseTarget.name = snapshotTitle;
                    if (signal.aborted) {
                        logger.warn('Command "check" was aborted after rendering');
                        throw new abort_error_1.AbortError('Command "check" was aborted after rendering');
                    }
                    else if (baseEyes.aborted) {
                        logger.warn(`Renderer with id ${baseEyes.test.rendererId} was aborted during one of the previous steps`);
                        throw new abort_error_1.AbortError(`Renderer with id "${baseEyes.test.rendererId}" was aborted during one of the previous steps`);
                    }
                    const [result] = await baseEyes.checkAndClose({
                        target: { ...baseTarget, isTransformed: true },
                        settings: baseSettings,
                        logger,
                    });
                    if (baseEyes.aborted) {
                        logger.warn(`Renderer with id ${baseEyes.test.rendererId} was aborted during one of the previous steps`);
                        throw new abort_error_1.AbortError(`Renderer with id "${baseEyes.test.rendererId}" was aborted during one of the previous steps`);
                    }
                    return { ...result, eyes, renderer };
                }
                catch (error) {
                    await baseEyes.abort();
                    error.info = { eyes: baseEyes };
                    throw error;
                }
            }
            catch (error) {
                error.info = { ...error.info, userTestId: eyes.test.userTestId, renderer };
                throw error;
            }
        });
        return Promise.all(promises.map(async (promise) => {
            var _a, _b;
            try {
                const result = await promise;
                return { ...result, userTestId: eyes.test.userTestId, asExpected: true };
            }
            catch (error) {
                await ((_b = (_a = error.info) === null || _a === void 0 ? void 0 : _a.eyes) === null || _b === void 0 ? void 0 : _b.abort({ logger }));
                throw error;
            }
        }));
    };
}
exports.makeCheckAndClose = makeCheckAndClose;
