// @applitools/core/dist/ufg/open-eyes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeOpenEyes = void 0;
const driver_1 = require("@applitools/driver");
const ufg_client_1 = require("@applitools/ufg-client");
const get_base_eyes_1 = require("./get-base-eyes");
const check_1 = require("./check");
const check_and_close_1 = require("./check-and-close");
const close_1 = require("./close");
const abort_1 = require("./abort");
const abort_controller_1 = require("abort-controller");
const utils = __importStar(require("@applitools/utils"));
function makeOpenEyes({ core, client, spec, logger: defaultLogger, }) {
    return async function openEyes({ target, settings, eyes, logger = defaultLogger, }) {
        var _a, _b;
        logger.log(`Command "openEyes" is called with ${target ? 'default driver and' : ''}`, ...(settings ? ['settings', settings] : []), eyes ? 'predefined eyes' : '');
        const driver = target && (await (0, driver_1.makeDriver)({ spec, driver: target, logger, customConfig: { disableHelper: true } }));
        if (driver && !eyes) {
            const currentContext = driver.currentContext;
            if ((_a = settings.environment) === null || _a === void 0 ? void 0 : _a.viewportSize) {
                await driver.setViewportSize(settings.environment.viewportSize);
            }
            await currentContext.focus();
        }
        const controller = new abort_controller_1.AbortController();
        const account = await core.getAccountInfo({ settings, logger });
        client !== null && client !== void 0 ? client : (client = (0, ufg_client_1.makeUFGClient)({
            config: { ...account.ufg, ...account, proxy: settings.proxy },
            concurrency: (_b = settings.renderConcurrency) !== null && _b !== void 0 ? _b : 5,
            logger,
        }));
        const getBaseEyes = (0, get_base_eyes_1.makeGetBaseEyes)({ settings, eyes, core, client, logger });
        return utils.general.extend({}, eyes => {
            var _a;
            const storage = [];
            let closed = false;
            let aborted = false;
            return {
                type: 'ufg',
                test: {
                    userTestId: settings.userTestId,
                    batchId: (_a = settings.batch) === null || _a === void 0 ? void 0 : _a.id,
                    keepBatchOpen: settings.keepBatchOpen,
                    server: { serverUrl: settings.serverUrl, apiKey: settings.apiKey, proxy: settings.proxy },
                    account,
                },
                get running() {
                    return !closed && !aborted;
                },
                get closed() {
                    return closed;
                },
                get aborted() {
                    return aborted;
                },
                getBaseEyes,
                // check with indexing and storage
                check: utils.general.wrap((0, check_1.makeCheck)({ eyes, client, target: driver, spec, signal: controller.signal, logger }), async (check, options = {}) => {
                    const results = await check(options);
                    storage.push(...results.map(result => ({ promise: result.promise, renderer: result.renderer })));
                    return results;
                }),
                checkAndClose: (0, check_and_close_1.makeCheckAndClose)({ eyes, client, target: driver, spec, signal: controller.signal, logger }),
                // close only once
                close: utils.general.wrap((0, close_1.makeClose)({ storage, logger }), async (close, options) => {
                    if (closed || aborted)
                        return [];
                    closed = true;
                    return close(options);
                }),
                // abort only once
                abort: utils.general.wrap((0, abort_1.makeAbort)({ storage, controller, logger }), async (abort, options) => {
                    if (aborted || closed)
                        return [];
                    aborted = true;
                    return abort(options);
                }),
            };
        });
    };
}
exports.makeOpenEyes = makeOpenEyes;
