// @superset-ui/core/lib/connection/constants.js
"use strict";

exports.__esModule = true;
exports.HTTP_STATUS_OK = exports.HTTP_STATUS_NOT_MODIFIED = exports.DEFAULT_FETCH_RETRY_OPTIONS = exports.DEFAULT_BASE_URL = exports.CACHE_KEY = exports.CACHE_AVAILABLE = void 0;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const DEFAULT_BASE_URL = 'http://localhost'; // HTTP status codes

exports.DEFAULT_BASE_URL = DEFAULT_BASE_URL;
const HTTP_STATUS_OK = 200;
exports.HTTP_STATUS_OK = HTTP_STATUS_OK;
const HTTP_STATUS_NOT_MODIFIED = 304; // Namespace for Cache API

exports.HTTP_STATUS_NOT_MODIFIED = HTTP_STATUS_NOT_MODIFIED;
const CACHE_AVAILABLE = ('caches' in window);
exports.CACHE_AVAILABLE = CACHE_AVAILABLE;
const CACHE_KEY = '@SUPERSET-UI/CONNECTION';
exports.CACHE_KEY = CACHE_KEY;
const DEFAULT_FETCH_RETRY_OPTIONS = {
  retries: 3,
  retryDelay: 1000,
  retryOn: [503]
};
exports.DEFAULT_FETCH_RETRY_OPTIONS = DEFAULT_FETCH_RETRY_OPTIONS;
