// @vx/glyph/build/glyphs/statistics/BoxPlot.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = GlyphBoxPlot;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Glyph = require('../Glyph');

var _Glyph2 = _interopRequireDefault(_Glyph);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function GlyphBoxPlot(_ref) {
  var _ref$left = _ref.left,
      left = _ref$left === undefined ? 0 : _ref$left,
      className = _ref.className,
      max = _ref.max,
      min = _ref.min,
      firstQuartile = _ref.firstQuartile,
      thirdQuartile = _ref.thirdQuartile,
      median = _ref.median,
      boxWidth = _ref.boxWidth,
      fill = _ref.fill,
      stroke = _ref.stroke,
      strokeWidth = _ref.strokeWidth;

  var centerX = left + boxWidth / 2;
  return _react2.default.createElement(
    _Glyph2.default,
    {
      className: (0, _classnames2.default)('vx-glyph-boxplot', className)
    },
    _react2.default.createElement('line', {
      x1: centerX - boxWidth / 4,
      y1: max,
      x2: centerX + boxWidth / 4,
      y2: max,
      stroke: stroke,
      strokeWidth: strokeWidth
    }),
    _react2.default.createElement('line', {
      x1: centerX,
      y1: max,
      x2: centerX,
      y2: thirdQuartile,
      stroke: stroke,
      strokeWidth: strokeWidth
    }),
    _react2.default.createElement('rect', {
      x: left,
      y: thirdQuartile,
      width: boxWidth,
      height: firstQuartile - thirdQuartile,
      stroke: stroke,
      strokeWidth: strokeWidth,
      fill: fill,
      rx: 2,
      ry: 2
    }),
    _react2.default.createElement('line', {
      x1: left,
      y1: median,
      x2: left + boxWidth,
      y2: median,
      stroke: stroke,
      strokeWidth: strokeWidth
    }),
    _react2.default.createElement('line', {
      x1: centerX,
      y1: firstQuartile,
      x2: centerX,
      y2: min,
      stroke: stroke,
      strokeWidth: strokeWidth
    }),
    _react2.default.createElement('line', {
      x1: centerX - boxWidth / 4,
      y1: min,
      x2: centerX + boxWidth / 4,
      y2: min,
      stroke: stroke,
      strokeWidth: strokeWidth
    })
  );
}
