// @applitools/driver/dist/user-agent-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUserAgentData = void 0;
const WINDOWS_VERSIONS = {
    '0.1.0': '7',
    '0.2.0': '8',
    '0.3.0': '8.1',
    '10.0.0': '10',
    '15.0.0': '11',
};
function parseUserAgentData(userAgentData) {
    var _a, _b, _c;
    const chromiumBrand = (_a = userAgentData.brands) === null || _a === void 0 ? void 0 : _a.find(brand => /Chromium/i.test(brand.brand));
    const browserBrand = (_c = (_b = userAgentData.brands) === null || _b === void 0 ? void 0 : _b.find(brand => brand !== chromiumBrand && !/Not.?A.?Brand/i.test(brand.brand))) !== null && _c !== void 0 ? _c : chromiumBrand;
    const info = {
        browserName: browserBrand === null || browserBrand === void 0 ? void 0 : browserBrand.brand,
        browserVersion: browserBrand === null || browserBrand === void 0 ? void 0 : browserBrand.version,
        platformName: userAgentData.platform || undefined,
        platformVersion: userAgentData.platformVersion || undefined,
        deviceName: userAgentData.model || undefined,
        isMobile: userAgentData.mobile,
        isChromium: Boolean(chromiumBrand),
    };
    if (info.platformName === 'Windows') {
        info.platformVersion = WINDOWS_VERSIONS[info.platformVersion];
    }
    else if (info.platformName === 'macOS') {
        info.platformName = 'Mac OS X';
        info.platformVersion = info.platformVersion.split(/[._]/, 2).join('.');
    }
    return info;
}
exports.parseUserAgentData = parseUserAgentData;
