// @applitools/core/dist/close.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeClose = void 0;
const test_error_1 = require("./errors/test-error");
function makeClose({ eyes, logger: defaultLogger, }) {
    return async function close({ settings, config, logger = defaultLogger, } = {}) {
        var _a;
        settings = { ...config === null || config === void 0 ? void 0 : config.close, ...settings };
        (_a = settings.updateBaselineIfNew) !== null && _a !== void 0 ? _a : (settings.updateBaselineIfNew = true);
        const typedEyes = await eyes.getTypedEyes({ logger });
        const results = await typedEyes.close({ settings, logger });
        if (settings.throwErr) {
            results.forEach(result => {
                if (result.status !== 'Passed')
                    throw new test_error_1.TestError(result);
            });
        }
        return results.length > 0
            ? results
            : [
                {
                    userTestId: eyes.test.userTestId,
                    name: '',
                    steps: 0,
                    matches: 0,
                    mismatches: 0,
                    missing: 0,
                    exactMatches: 0,
                    strictMatches: 0,
                    contentMatches: 0,
                    layoutMatches: 0,
                    noneMatches: 0,
                },
            ];
    };
}
exports.makeClose = makeClose;
