// @applitools/core/dist/core.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCore = void 0;
const logger_1 = require("@applitools/logger");
const core_base_1 = require("@applitools/core-base");
const get_viewport_size_1 = require("./automation/get-viewport-size");
const set_viewport_size_1 = require("./automation/set-viewport-size");
const locate_1 = require("./locate");
const open_eyes_1 = require("./open-eyes");
const make_manager_1 = require("./make-manager");
const close_batch_1 = require("./close-batch");
const delete_test_1 = require("./delete-test");
const utils = __importStar(require("@applitools/utils"));
function makeCore({ spec, concurrency, core, agentId = 'core', cwd = process.cwd(), logger, } = {}) {
    var _a;
    logger = (_a = logger === null || logger === void 0 ? void 0 : logger.extend({ label: 'core' })) !== null && _a !== void 0 ? _a : (0, logger_1.makeLogger)({ label: 'core' });
    logger.log(`Core is initialized ${core ? 'with' : 'without'} custom base core`);
    core !== null && core !== void 0 ? core : (core = (0, core_base_1.makeCore)({ agentId, cwd, logger }));
    return utils.general.extend(core, {
        isDriver: spec === null || spec === void 0 ? void 0 : spec.isDriver,
        isElement: spec === null || spec === void 0 ? void 0 : spec.isElement,
        isSelector: spec === null || spec === void 0 ? void 0 : spec.isSelector,
        getViewportSize: (0, get_viewport_size_1.makeGetViewportSize)({ spec, logger }),
        setViewportSize: (0, set_viewport_size_1.makeSetViewportSize)({ spec, logger }),
        locate: (0, locate_1.makeLocate)({ spec, core, logger }),
        openEyes: (0, open_eyes_1.makeOpenEyes)({ spec, core, concurrency, logger }),
        makeManager: (0, make_manager_1.makeMakeManager)({ spec, concurrency, agentId, logger }),
        closeBatch: (0, close_batch_1.makeCloseBatch)({ core, logger }),
        deleteTest: (0, delete_test_1.makeDeleteTest)({ core, logger }),
    });
}
exports.makeCore = makeCore;
