// @superset-ui/plugin-chart-echarts/lib/Timeseries/EchartsTimeseries.js
"use strict";

exports.__esModule = true;
exports.default = EchartsTimeseries;

var _react = _interopRequireWildcard(require("react"));

var _Echart = _interopRequireDefault(require("../components/Echart"));

var _series = require("../utils/series");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const TIMER_DURATION = 300; // @ts-ignore

function EchartsTimeseries({
  formData,
  height,
  width,
  echartOptions,
  groupby,
  labelMap,
  selectedValues,
  setDataMask,
  legendData = []
}) {
  const {
    emitFilter,
    stack
  } = formData;
  const echartRef = (0, _react.useRef)(null);
  const lastTimeRef = (0, _react.useRef)(Date.now());
  const lastSelectedLegend = (0, _react.useRef)('');
  const clickTimer = (0, _react.useRef)();
  const handleDoubleClickChange = (0, _react.useCallback)(name => {
    var _echartRef$current;

    const echartInstance = (_echartRef$current = echartRef.current) == null ? void 0 : _echartRef$current.getEchartInstance();

    if (!name) {
      _series.currentSeries.legend = '';
      echartInstance == null ? void 0 : echartInstance.dispatchAction({
        type: 'legendAllSelect'
      });
    } else {
      legendData.forEach(datum => {
        if (datum === name) {
          _series.currentSeries.legend = datum;
          echartInstance == null ? void 0 : echartInstance.dispatchAction({
            type: 'legendSelect',
            name: datum
          });
        } else {
          echartInstance == null ? void 0 : echartInstance.dispatchAction({
            type: 'legendUnSelect',
            name: datum
          });
        }
      });
    }
  }, [legendData]);

  const getModelInfo = (target, globalModel) => {
    let el = target;
    let model = null;

    while (el) {
      // eslint-disable-next-line no-underscore-dangle
      const modelInfo = el.__ecComponentInfo;

      if (modelInfo != null) {
        model = globalModel.getComponent(modelInfo.mainType, modelInfo.index);
        break;
      }

      el = el.parent;
    }

    return model;
  };

  const handleChange = (0, _react.useCallback)(values => {
    if (!emitFilter) {
      return;
    }

    const groupbyValues = values.map(value => labelMap[value]);
    setDataMask({
      extraFormData: {
        filters: values.length === 0 ? [] : groupby.map((col, idx) => {
          const val = groupbyValues.map(v => v[idx]);
          if (val === null || val === undefined) return {
            col,
            op: 'IS NULL'
          };
          return {
            col,
            op: 'IN',
            val: val
          };
        })
      },
      filterState: {
        label: groupbyValues.length ? groupbyValues : undefined,
        value: groupbyValues.length ? groupbyValues : null,
        selectedValues: values.length ? values : null
      }
    });
  }, [groupby, labelMap, setDataMask]);
  const eventHandlers = {
    click: props => {
      if (clickTimer.current) {
        clearTimeout(clickTimer.current);
      } // Ensure that double-click events do not trigger single click event. So we put it in the timer.


      clickTimer.current = setTimeout(() => {
        const {
          seriesName: name
        } = props;
        const values = Object.values(selectedValues);

        if (values.includes(name)) {
          handleChange(values.filter(v => v !== name));
        } else {
          handleChange([name]);
        }
      }, TIMER_DURATION);
    },
    mouseout: () => {
      _series.currentSeries.name = '';
    },
    mouseover: params => {
      _series.currentSeries.name = params.seriesName;
    },
    legendselectchanged: payload => {
      const currentTime = Date.now(); // TIMER_DURATION is the interval between two legendselectchanged event

      if (currentTime - lastTimeRef.current < TIMER_DURATION && lastSelectedLegend.current === payload.name) {
        // execute dbclick
        handleDoubleClickChange(payload.name);
      } else {
        lastTimeRef.current = currentTime; // remember last selected legend

        lastSelectedLegend.current = payload.name;
      } // if all legend is unselected, we keep all selected


      if (Object.values(payload.selected).every(i => !i)) {
        handleDoubleClickChange();
      }
    }
  };
  const zrEventHandlers = {
    dblclick: params => {
      var _echartRef$current2;

      // clear single click timer
      if (clickTimer.current) {
        clearTimeout(clickTimer.current);
      }

      const pointInPixel = [params.offsetX, params.offsetY];
      const echartInstance = (_echartRef$current2 = echartRef.current) == null ? void 0 : _echartRef$current2.getEchartInstance();

      if (echartInstance != null && echartInstance.containPixel('grid', pointInPixel)) {
        var _params$target;

        // do not trigger if click unstacked chart's blank area
        if (!stack && ((_params$target = params.target) == null ? void 0 : _params$target.type) === 'ec-polygon') return; // @ts-ignore

        const globalModel = echartInstance.getModel();
        const model = getModelInfo(params.target, globalModel);
        const seriesCount = globalModel.getSeriesCount();
        const currentSeriesIndices = globalModel.getCurrentSeriesIndices();

        if (model) {
          const {
            name
          } = model;

          if (seriesCount !== currentSeriesIndices.length) {
            handleDoubleClickChange();
          } else {
            handleDoubleClickChange(name);
          }
        }
      }
    }
  };
  return (0, _react2.jsx)(_Echart.default, {
    ref: echartRef,
    height: height,
    width: width,
    echartOptions: echartOptions,
    eventHandlers: eventHandlers,
    zrEventHandlers: zrEventHandlers,
    selectedValues: selectedValues
  });
}
