// @data-ui/sparkline/lib/annotation/Label.js
"use strict";

exports.__esModule = true;
exports.default = exports.defaultProps = exports.propTypes = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _theme = require("@data-ui/theme");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var propTypes = {
  anchor: _propTypes.default.oneOf(['start', 'middle', 'end']),
  dominantBaseline: _propTypes.default.string,
  dx: _propTypes.default.number,
  dy: _propTypes.default.number,
  fill: _propTypes.default.string,
  label: _propTypes.default.node,
  paintOrder: _propTypes.default.string,
  stroke: _propTypes.default.string,
  strokeWidth: _propTypes.default.number,
  x: _propTypes.default.number,
  y: _propTypes.default.number
};
exports.propTypes = propTypes;
var defaultProps = {
  anchor: 'middle',
  dominantBaseline: 'middle',
  dx: 0,
  dy: 0,
  fill: _theme.color.text,
  label: null,
  paintOrder: 'stroke',
  stroke: '#fff',
  strokeWidth: 2,
  x: 0,
  y: 0
};
exports.defaultProps = defaultProps;

function Label(_ref) {
  var anchor = _ref.anchor,
      dominantBaseline = _ref.dominantBaseline,
      dx = _ref.dx,
      dy = _ref.dy,
      fill = _ref.fill,
      label = _ref.label,
      paintOrder = _ref.paintOrder,
      stroke = _ref.stroke,
      strokeWidth = _ref.strokeWidth,
      x = _ref.x,
      y = _ref.y,
      rest = _objectWithoutPropertiesLoose(_ref, ["anchor", "dominantBaseline", "dx", "dy", "fill", "label", "paintOrder", "stroke", "strokeWidth", "x", "y"]);

  return _react.default.createElement("text", _extends({
    x: x,
    y: y,
    dx: dx,
    dy: dy,
    fill: fill,
    dominantBaseline: dominantBaseline,
    paintOrder: paintOrder,
    stroke: stroke,
    strokeWidth: strokeWidth,
    textAnchor: anchor
  }, rest), label);
}

Label.propTypes = propTypes;
Label.defaultProps = defaultProps;
Label.displayName = 'Label';
var _default = Label;
exports.default = _default;
