// @applitools/core/dist/utils/wait-for-lazy-load.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForLazyLoad = void 0;
const snippets_1 = require("@applitools/snippets");
async function waitForLazyLoad({ context, settings, logger, }) {
    var _a, _b, _c, _d, _e;
    logger.log('lazy loading the page before capturing a screenshot');
    await context.executePoll(snippets_1.lazyLoad, {
        main: [
            await context.getScrollingElement(),
            {
                scrollLength: (_a = settings.scrollLength) !== null && _a !== void 0 ? _a : 300,
                waitingTime: (_b = settings.waitingTime) !== null && _b !== void 0 ? _b : 2000,
                maxAmountToScroll: (_c = settings.maxAmountToScroll) !== null && _c !== void 0 ? _c : 15000,
            },
        ],
        poll: [],
        executionTimeout: 5 * 60 * 1000,
        pollTimeout: (_e = (_d = settings.pollTimeout) !== null && _d !== void 0 ? _d : settings.waitingTime) !== null && _e !== void 0 ? _e : 2000,
    });
}
exports.waitForLazyLoad = waitForLazyLoad;
