// @superset-ui/plugin-chart-table/lib/buildQuery.js
"use strict";

exports.__esModule = true;
exports.default = exports.cachedBuildQuery = void 0;
exports.getQueryMode = getQueryMode;

var _core = require("@superset-ui/core");

var _externalAPIs = require("./DataTable/utils/externalAPIs");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * Infer query mode from form data. If `all_columns` is set, then raw records mode,
 * otherwise defaults to aggregation mode.
 *
 * The same logic is used in `controlPanel` with control values as well.
 */
function getQueryMode(formData) {
  const {
    query_mode: mode
  } = formData;

  if (mode === _core.QueryMode.aggregate || mode === _core.QueryMode.raw) {
    return mode;
  }

  const rawColumns = formData == null ? void 0 : formData.all_columns;
  const hasRawColumns = rawColumns && rawColumns.length > 0;
  return hasRawColumns ? _core.QueryMode.raw : _core.QueryMode.aggregate;
}

const buildQuery = (formData, options) => {
  const {
    percent_metrics: percentMetrics,
    order_desc: orderDesc = false
  } = formData;
  const queryMode = getQueryMode(formData);
  const sortByMetric = (0, _core.ensureIsArray)(formData.timeseries_limit_metric)[0];
  let formDataCopy = formData; // never include time in raw records mode

  if (queryMode === _core.QueryMode.raw) {
    formDataCopy = { ...formData,
      include_time: false
    };
  }

  return (0, _core.buildQueryContext)(formDataCopy, baseQueryObject => {
    var _options$extras, _options$extras$cache, _options$extras2, _options$extras2$cach, _options$hooks2, _metrics2, _formData$extra_form_;

    let {
      metrics,
      orderby = []
    } = baseQueryObject;
    let postProcessing = [];

    if (queryMode === _core.QueryMode.aggregate) {
      var _metrics;

      metrics = metrics || []; // orverride orderby with timeseries metric when in aggregation mode

      if (sortByMetric) {
        orderby = [[sortByMetric, !orderDesc]];
      } else if (((_metrics = metrics) == null ? void 0 : _metrics.length) > 0) {
        // default to ordering by first metric in descending order
        // when no "sort by" metric is set (regargless if "SORT DESC" is set to true)
        orderby = [[metrics[0], false]];
      } // add postprocessing for percent metrics only when in aggregation mode


      if (percentMetrics && percentMetrics.length > 0) {
        const percentMetricLabels = (0, _core.removeDuplicates)(percentMetrics.map(_core.getMetricLabel));
        metrics = (0, _core.removeDuplicates)(metrics.concat(percentMetrics), _core.getMetricLabel);
        postProcessing = [{
          operation: 'contribution',
          options: {
            columns: percentMetricLabels,
            rename_columns: percentMetricLabels.map(x => `%${x}`)
          }
        }];
      }
    }

    const moreProps = {};
    const ownState = (options == null ? void 0 : options.ownState) ?? {};

    if (formDataCopy.server_pagination) {
      moreProps.row_limit = ownState.pageSize ?? formDataCopy.server_page_length;
      moreProps.row_offset = (ownState.currentPage ?? 0) * (ownState.pageSize ?? 0);
    }

    let queryObject = { ...baseQueryObject,
      orderby,
      metrics,
      post_processing: postProcessing,
      ...moreProps
    };

    if (formData.server_pagination && options != null && (_options$extras = options.extras) != null && (_options$extras$cache = _options$extras.cachedChanges) != null && _options$extras$cache[formData.slice_id] && JSON.stringify(options == null ? void 0 : (_options$extras2 = options.extras) == null ? void 0 : (_options$extras2$cach = _options$extras2.cachedChanges) == null ? void 0 : _options$extras2$cach[formData.slice_id]) !== JSON.stringify(queryObject.filters)) {
      var _options$hooks;

      queryObject = { ...queryObject,
        row_offset: 0
      };
      (0, _externalAPIs.updateExternalFormData)(options == null ? void 0 : (_options$hooks = options.hooks) == null ? void 0 : _options$hooks.setDataMask, 0, queryObject.row_limit ?? 0);
    } // Because we use same buildQuery for all table on the page we need split them by id


    options == null ? void 0 : (_options$hooks2 = options.hooks) == null ? void 0 : _options$hooks2.setCachedChanges({
      [formData.slice_id]: queryObject.filters
    });
    const extraQueries = [];

    if ((_metrics2 = metrics) != null && _metrics2.length && formData.show_totals && queryMode === _core.QueryMode.aggregate) {
      extraQueries.push({ ...queryObject,
        columns: [],
        row_limit: 0,
        row_offset: 0,
        post_processing: []
      });
    }

    const interactiveGroupBy = (_formData$extra_form_ = formData.extra_form_data) == null ? void 0 : _formData$extra_form_.interactive_groupby;

    if (interactiveGroupBy && queryObject.columns) {
      queryObject.columns = [...new Set([...queryObject.columns, ...interactiveGroupBy])];
    }

    if (formData.server_pagination) {
      return [{ ...queryObject
      }, { ...queryObject,
        row_limit: 0,
        row_offset: 0,
        post_processing: [],
        is_rowcount: true
      }, ...extraQueries];
    }

    return [queryObject, ...extraQueries];
  });
}; // Use this closure to cache changing of external filters, if we have server pagination we need reset page to 0, after
// external filter changed


const cachedBuildQuery = () => {
  let cachedChanges = {};

  const setCachedChanges = newChanges => {
    cachedChanges = { ...cachedChanges,
      ...newChanges
    };
  };

  return (formData, options) => buildQuery({ ...formData
  }, {
    extras: {
      cachedChanges
    },
    ownState: (options == null ? void 0 : options.ownState) ?? {},
    hooks: { ...(options == null ? void 0 : options.hooks),
      setDataMask: () => {},
      setCachedChanges
    }
  });
};

exports.cachedBuildQuery = cachedBuildQuery;

var _default = cachedBuildQuery();

exports.default = _default;
