// @superset-ui/plugin-chart-table/lib/DataTable/DataTable.js
"use strict";

exports.__esModule = true;
exports.default = DataTable;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _reactTable = require("react-table");

var _matchSorter = require("match-sorter");

var _GlobalFilter = _interopRequireDefault(require("./components/GlobalFilter"));

var _SelectPageSize = _interopRequireDefault(require("./components/SelectPageSize"));

var _Pagination = _interopRequireDefault(require("./components/Pagination"));

var _useSticky = _interopRequireDefault(require("./hooks/useSticky"));

var _consts = require("../consts");

var _react2 = require("@emotion/react");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

// Be sure to pass our updateMyData and the skipReset option
function DataTable({
  tableClassName,
  columns,
  data,
  serverPaginationData,
  width: initialWidth = '100%',
  height: initialHeight = 300,
  pageSize: initialPageSize = 0,
  initialState: initialState_ = {},
  pageSizeOptions = _consts.PAGE_SIZE_OPTIONS,
  maxPageItemCount = 9,
  sticky: doSticky,
  searchInput = true,
  onServerPaginationChange,
  rowCount,
  selectPageSize,
  noResults: noResultsText = 'No data found',
  hooks,
  serverPagination,
  wrapperRef: userWrapperRef,
  ...moreUseTableOptions
}) {
  const tableHooks = [_reactTable.useGlobalFilter, _reactTable.useSortBy, _reactTable.usePagination, doSticky ? _useSticky.default : [], hooks || []].flat();
  const resultsSize = serverPagination ? rowCount : data.length;
  const sortByRef = (0, _react.useRef)([]); // cache initial `sortby` so sorting doesn't trigger page reset

  const pageSizeRef = (0, _react.useRef)([initialPageSize, resultsSize]);
  const hasPagination = initialPageSize > 0 && resultsSize > 0; // pageSize == 0 means no pagination

  const hasGlobalControl = hasPagination || !!searchInput;
  const initialState = { ...initialState_,
    // zero length means all pages, the `usePagination` plugin does not
    // understand pageSize = 0
    sortBy: sortByRef.current,
    pageSize: initialPageSize > 0 ? initialPageSize : resultsSize || 10
  };
  const defaultWrapperRef = (0, _react.useRef)(null);
  const globalControlRef = (0, _react.useRef)(null);
  const paginationRef = (0, _react.useRef)(null);
  const wrapperRef = userWrapperRef || defaultWrapperRef;
  const paginationData = JSON.stringify(serverPaginationData);
  const defaultGetTableSize = (0, _react.useCallback)(() => {
    if (wrapperRef.current) {
      var _globalControlRef$cur, _paginationRef$curren;

      // `initialWidth` and `initialHeight` could be also parameters like `100%`
      // `Number` reaturns `NaN` on them, then we fallback to computed size
      const width = Number(initialWidth) || wrapperRef.current.clientWidth;
      const height = (Number(initialHeight) || wrapperRef.current.clientHeight) - (((_globalControlRef$cur = globalControlRef.current) == null ? void 0 : _globalControlRef$cur.clientHeight) || 0) - (((_paginationRef$curren = paginationRef.current) == null ? void 0 : _paginationRef$curren.clientHeight) || 0);
      return {
        width,
        height
      };
    }

    return undefined; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialHeight, initialWidth, wrapperRef, hasPagination, hasGlobalControl, paginationRef, resultsSize, paginationData]);
  const defaultGlobalFilter = (0, _react.useCallback)((rows, columnIds, filterValue) => {
    // allow searching by "col1_value col2_value"
    const joinedString = row => columnIds.map(x => row.values[x]).join(' ');

    return (0, _matchSorter.matchSorter)(rows, filterValue, {
      keys: [...columnIds, joinedString],
      threshold: _matchSorter.rankings.ACRONYM
    });
  }, []);
  const {
    getTableProps,
    getTableBodyProps,
    prepareRow,
    headerGroups,
    footerGroups,
    page,
    pageCount,
    gotoPage,
    preGlobalFilteredRows,
    setGlobalFilter,
    setPageSize: setPageSize_,
    wrapStickyTable,
    state: {
      pageIndex,
      pageSize,
      globalFilter: filterValue,
      sticky = {}
    }
  } = (0, _reactTable.useTable)({
    columns,
    data,
    initialState,
    getTableSize: defaultGetTableSize,
    globalFilter: defaultGlobalFilter,
    ...moreUseTableOptions
  }, ...tableHooks); // make setPageSize accept 0

  const setPageSize = size => {
    if (serverPagination) {
      onServerPaginationChange(0, size);
    } // keep the original size if data is empty


    if (size || resultsSize !== 0) {
      setPageSize_(size === 0 ? resultsSize : size);
    }
  };

  const noResults = typeof noResultsText === 'function' ? noResultsText(filterValue) : noResultsText;

  const getNoResults = () => (0, _react2.jsx)("div", {
    className: "dt-no-results"
  }, noResults);

  if (!columns || columns.length === 0) {
    return wrapStickyTable ? wrapStickyTable(getNoResults) : getNoResults();
  }

  const shouldRenderFooter = columns.some(x => !!x.Footer);

  const renderTable = () => (0, _react2.jsx)("table", getTableProps({
    className: tableClassName
  }), (0, _react2.jsx)("thead", null, headerGroups.map(headerGroup => {
    const {
      key: headerGroupKey,
      ...headerGroupProps
    } = headerGroup.getHeaderGroupProps();
    return (0, _react2.jsx)("tr", _extends({
      key: headerGroupKey || headerGroup.id
    }, headerGroupProps), headerGroup.headers.map(column => column.render('Header', {
      key: column.id,
      ...column.getSortByToggleProps()
    })));
  })), (0, _react2.jsx)("tbody", getTableBodyProps(), page && page.length > 0 ? page.map(row => {
    prepareRow(row);
    const {
      key: rowKey,
      ...rowProps
    } = row.getRowProps();
    return (0, _react2.jsx)("tr", _extends({
      key: rowKey || row.id
    }, rowProps), row.cells.map(cell => cell.render('Cell', {
      key: cell.column.id
    })));
  }) : (0, _react2.jsx)("tr", null, (0, _react2.jsx)("td", {
    className: "dt-no-results",
    colSpan: columns.length
  }, noResults))), shouldRenderFooter && (0, _react2.jsx)("tfoot", null, footerGroups.map(footerGroup => {
    const {
      key: footerGroupKey,
      ...footerGroupProps
    } = footerGroup.getHeaderGroupProps();
    return (0, _react2.jsx)("tr", _extends({
      key: footerGroupKey || footerGroup.id
    }, footerGroupProps), footerGroup.headers.map(column => column.render('Footer', {
      key: column.id
    })));
  }))); // force update the pageSize when it's been update from the initial state


  if (pageSizeRef.current[0] !== initialPageSize || // when initialPageSize stays as zero, but total number of records changed,
  // we'd also need to update page size
  initialPageSize === 0 && pageSizeRef.current[1] !== resultsSize) {
    pageSizeRef.current = [initialPageSize, resultsSize];
    setPageSize(initialPageSize);
  }

  const paginationStyle = sticky.height ? {} : {
    visibility: 'hidden'
  };
  let resultPageCount = pageCount;
  let resultCurrentPageSize = pageSize;
  let resultCurrentPage = pageIndex;
  let resultOnPageChange = gotoPage;

  if (serverPagination) {
    const serverPageSize = serverPaginationData.pageSize ?? initialPageSize;
    resultPageCount = Math.ceil(rowCount / serverPageSize);

    if (!Number.isFinite(resultPageCount)) {
      resultPageCount = 0;
    }

    resultCurrentPageSize = serverPageSize;
    const foundPageSizeIndex = pageSizeOptions.findIndex(([option]) => option >= resultCurrentPageSize);

    if (foundPageSizeIndex === -1) {
      resultCurrentPageSize = 0;
    }

    resultCurrentPage = serverPaginationData.currentPage ?? 0;

    resultOnPageChange = pageNumber => onServerPaginationChange(pageNumber, serverPageSize);
  }

  return (0, _react2.jsx)("div", {
    ref: wrapperRef,
    style: {
      width: initialWidth,
      height: initialHeight
    }
  }, hasGlobalControl ? (0, _react2.jsx)("div", {
    ref: globalControlRef,
    className: "form-inline dt-controls"
  }, (0, _react2.jsx)("div", {
    className: "row"
  }, (0, _react2.jsx)("div", {
    className: "col-sm-6"
  }, hasPagination ? (0, _react2.jsx)(_SelectPageSize.default, {
    total: resultsSize,
    current: resultCurrentPageSize,
    options: pageSizeOptions,
    selectRenderer: typeof selectPageSize === 'boolean' ? undefined : selectPageSize,
    onChange: setPageSize
  }) : null), searchInput ? (0, _react2.jsx)("div", {
    className: "col-sm-6"
  }, (0, _react2.jsx)(_GlobalFilter.default, {
    searchInput: typeof searchInput === 'boolean' ? undefined : searchInput,
    preGlobalFilteredRows: preGlobalFilteredRows,
    setGlobalFilter: setGlobalFilter,
    filterValue: filterValue
  })) : null)) : null, wrapStickyTable ? wrapStickyTable(renderTable) : renderTable(), hasPagination && resultPageCount > 1 ? (0, _react2.jsx)(_Pagination.default, {
    ref: paginationRef,
    style: paginationStyle,
    maxPageItemCount: maxPageItemCount,
    pageCount: resultPageCount,
    currentPage: resultCurrentPage,
    onPageChange: resultOnPageChange
  }) : null);
}

DataTable.propTypes = {
  tableClassName: _propTypes.default.string,
  pageSizeOptions: _propTypes.default.array,
  maxPageItemCount: _propTypes.default.number,
  hooks: _propTypes.default.array,
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  serverPagination: _propTypes.default.bool,
  onServerPaginationChange: _propTypes.default.func.isRequired,
  serverPaginationData: _propTypes.default.shape({
    pageSize: _propTypes.default.number,
    currentPage: _propTypes.default.number
  }).isRequired,
  pageSize: _propTypes.default.number,
  noResults: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
  sticky: _propTypes.default.bool,
  rowCount: _propTypes.default.number.isRequired
};
