// encodable/lib/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  Encoder: true,
  ChannelEncoder: true,
  createEncoderFactory: true,
  completeChannelDef: true,
  createScale: true,
  mergeEncoding: true,
  Encodable: true,
  isFieldDef: true,
  isNonValueDef: true,
  isTypedFieldDef: true,
  isValueDef: true,
  isScaleFieldDef: true,
  isPositionFieldDef: true,
  isCompleteFieldDef: true,
  isCompleteValueDef: true
};
exports.isCompleteValueDef = exports.isCompleteFieldDef = exports.isPositionFieldDef = exports.isScaleFieldDef = exports.isValueDef = exports.isTypedFieldDef = exports.isNonValueDef = exports.isFieldDef = exports.Encodable = exports.mergeEncoding = exports.createScale = exports.completeChannelDef = exports.createEncoderFactory = exports.ChannelEncoder = exports.Encoder = void 0;

var _Encoder = _interopRequireDefault(require("./encoders/Encoder"));

exports.Encoder = _Encoder.default;

var _ChannelEncoder = _interopRequireDefault(require("./encoders/ChannelEncoder"));

exports.ChannelEncoder = _ChannelEncoder.default;

var _createEncoderFactory = _interopRequireDefault(require("./encoders/createEncoderFactory"));

exports.createEncoderFactory = _createEncoderFactory.default;

var _completeChannelDef = _interopRequireDefault(require("./fillers/completeChannelDef"));

exports.completeChannelDef = _completeChannelDef.default;

var _createScale = _interopRequireDefault(require("./parsers/scale/createScale"));

exports.createScale = _createScale.default;

var _mergeEncoding = _interopRequireDefault(require("./utils/mergeEncoding"));

exports.mergeEncoding = _mergeEncoding.default;

var _Encodable = _interopRequireDefault(require("./options/Encodable"));

exports.Encodable = _Encodable.default;

var _resolvers = require("./options/resolvers");

Object.keys(_resolvers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _resolvers[key];
});

var _ChannelDef = require("./typeGuards/ChannelDef");

exports.isFieldDef = _ChannelDef.isFieldDef;
exports.isNonValueDef = _ChannelDef.isNonValueDef;
exports.isTypedFieldDef = _ChannelDef.isTypedFieldDef;
exports.isValueDef = _ChannelDef.isValueDef;
exports.isScaleFieldDef = _ChannelDef.isScaleFieldDef;
exports.isPositionFieldDef = _ChannelDef.isPositionFieldDef;

var _CompleteChannelDef = require("./typeGuards/CompleteChannelDef");

exports.isCompleteFieldDef = _CompleteChannelDef.isCompleteFieldDef;
exports.isCompleteValueDef = _CompleteChannelDef.isCompleteValueDef;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _types[key];
});

var _format = require("@encodable/format");

Object.keys(_format).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _format[key];
});

var _color = require("@encodable/color");

Object.keys(_color).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _color[key];
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
