// @superset-ui/preset-chart-xy/lib/components/legend/DefaultLegendGroup.js
"use strict";

exports.__esModule = true;
exports.default = DefaultLegendGroupRenderer;

var _react = _interopRequireDefault(require("react"));

var _DefaultLegendItem = _interopRequireDefault(require("./DefaultLegendItem"));

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const LEGEND_GROUP_STYLE = {
  display: 'flex',
  flexBasis: 'auto',
  flexDirection: 'row',
  flexGrow: 1,
  flexShrink: 1,
  flexWrap: 'wrap',
  fontSize: '0.8em',
  justifyContent: 'flex-end',
  padding: 8
};

function DefaultLegendGroupRenderer({
  group,
  ItemRenderer = _DefaultLegendItem.default,
  ItemMarkRenderer,
  ItemLabelRenderer,
  style
}) {
  const combinedStyle = typeof style === 'undefined' ? LEGEND_GROUP_STYLE : { ...LEGEND_GROUP_STYLE,
    ...style
  };
  return (0, _react2.jsx)("div", {
    style: combinedStyle
  }, 'items' in group && group.items.map(item => (0, _react2.jsx)(ItemRenderer, {
    key: `legend-item-${group.field}-${item.input}`,
    group: group,
    item: item,
    MarkRenderer: ItemMarkRenderer,
    LabelRenderer: ItemLabelRenderer
  })));
}
