// @superset-ui/plugin-chart-echarts/lib/Funnel/types.js
"use strict";

exports.__esModule = true;
exports.EchartsFunnelLabelTypeType = exports.DEFAULT_FORM_DATA = void 0;

var _types = require("../types");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
let EchartsFunnelLabelTypeType;
exports.EchartsFunnelLabelTypeType = EchartsFunnelLabelTypeType;

(function (EchartsFunnelLabelTypeType) {
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["Key"] = 0] = "Key";
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["Value"] = 1] = "Value";
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["Percent"] = 2] = "Percent";
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["KeyValue"] = 3] = "KeyValue";
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["KeyPercent"] = 4] = "KeyPercent";
  EchartsFunnelLabelTypeType[EchartsFunnelLabelTypeType["KeyValuePercent"] = 5] = "KeyValuePercent";
})(EchartsFunnelLabelTypeType || (exports.EchartsFunnelLabelTypeType = EchartsFunnelLabelTypeType = {}));

// @ts-ignore
const DEFAULT_FORM_DATA = { ..._types.DEFAULT_LEGEND_FORM_DATA,
  groupby: [],
  labelLine: false,
  labelType: EchartsFunnelLabelTypeType.Key,
  legendOrientation: _types.LegendOrientation.Top,
  legendType: _types.LegendType.Scroll,
  numberFormat: 'SMART_NUMBER',
  showLabels: true,
  sort: 'descending',
  orient: 'vertical',
  gap: 0,
  emitFilter: false
};
exports.DEFAULT_FORM_DATA = DEFAULT_FORM_DATA;
