// @superset-ui/legacy-plugin-chart-sankey-loop/lib/SankeyLoop.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Selection = require("d3-selection");

var _d3SankeyDiagram = require("d3-sankey-diagram");

var _core = require("@superset-ui/core");

require("./SankeyLoop.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/sort-prop-types */
// a problem with 'd3-sankey-diagram'  is that the sankey().extent() paramters, which
// informs the layout of the bounding box of the sankey columns, does not account
// for labels and paths which happen to be layedout outside that rectangle.
// for that reason i've selected relatively large default left/right margins, and have
// made 'margin' a property.   i have raised an issue in the chart repo:
//
//   https://github.com/ricklupton/d3-sankey-diagram/issues/20
const defaultMargin = {
  top: 0,
  right: 80,
  bottom: 0,
  left: 80
};
const propTypes = {
  data: _propTypes.default.arrayOf(_propTypes.default.shape({
    source: _propTypes.default.string,
    target: _propTypes.default.string,
    value: _propTypes.default.number
  })),
  width: _propTypes.default.number,
  height: _propTypes.default.number,
  colorScheme: _propTypes.default.string,
  margin: _propTypes.default.shape({
    top: _propTypes.default.number,
    right: _propTypes.default.number,
    bottom: _propTypes.default.number,
    left: _propTypes.default.number
  })
};
const percentFormat = (0, _core.getNumberFormatter)(_core.NumberFormats.PERCENT_1_POINT);
const countFormat = (0, _core.getNumberFormatter)();

function computeGraph(links) {
  // this assumes source and target are string values
  const nodes = Array.from(links.reduce((set, {
    source,
    target
  }) => set.add(source).add(target), new Set())).map(id => ({
    id,
    name: id
  }));
  return {
    nodes,
    // links are shallow copied as the chart layout modifies them, and it is best to
    // leave the passed data un-altered
    links: links.map(d => ({ ...d
    }))
  };
}

function SankeyLoop(element, props) {
  const {
    data,
    width,
    height,
    colorScheme
  } = props;

  const color = _core.CategoricalColorNamespace.getScale(colorScheme);

  const margin = { ...defaultMargin,
    ...props.margin
  };
  const innerWidth = width - margin.left - margin.right;
  const innerHeight = height - margin.top - margin.bottom;
  const layout = (0, _d3SankeyDiagram.sankey)().nodeId(d => d.id).extent([[margin.left, margin.top], [innerWidth, innerHeight]]);
  const diagram = (0, _d3SankeyDiagram.sankeyDiagram)().nodeTitle(d => d.name).linkTitle(({
    source: {
      name: sName,
      value: sValue
    },
    target: {
      name: tName
    },
    value
  }) => `${sName} → ${tName}: ${countFormat(value)} (${percentFormat(value / sValue)})`).linkColor(d => color(d.source.name));
  const div = (0, _d3Selection.select)(element);
  div.selectAll('*').remove();
  const svg = div.append('svg').classed('superset-legacy-chart-sankey-loop', true).style('width', width).style('height', height).datum(layout(computeGraph(data))).call(diagram);
  svg.selectAll('g.link').classed('link', true).append('text').attr('x', d => d.points[0].x).attr('y', d => d.points[0].y).attr('dy', 3).attr('dx', 2).text(d => `${countFormat(d.value)} (${percentFormat(d.value / d.source.value)})`);
}

SankeyLoop.displayName = 'SankeyLoop';
SankeyLoop.propTypes = propTypes;
var _default = SankeyLoop;
exports.default = _default;
