// @superset-ui/legacy-plugin-chart-event-flow/lib/EventFlow.js
"use strict";

exports.__esModule = true;
exports.default = EventFlow;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _eventFlow = require("@data-ui/event-flow");

var _core = require("@superset-ui/core");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
function EventFlow({
  data,
  initialMinEventCount,
  height = 400,
  width = 400
}) {
  if (data) {
    return (0, _react2.jsx)(_eventFlow.App, {
      width: width,
      height: height,
      data: data,
      initialMinEventCount: initialMinEventCount,
      initialShowControls: false
    });
  }

  return (0, _react2.jsx)("div", {
    style: {
      height,
      width
    }
  }, (0, _react2.jsx)("div", null, (0, _core.t)('Sorry, there appears to be no data')));
}

EventFlow.propTypes = {
  data: _propTypes.default.array.isRequired,
  height: _propTypes.default.number.isRequired,
  width: _propTypes.default.number.isRequired,
  initialMinEventCount: _propTypes.default.number.isRequired
};
