// @superset-ui/core/lib/query/types/AnnotationLayer.js
"use strict";

exports.__esModule = true;
exports.AnnotationType = exports.AnnotationStyle = exports.AnnotationSourceType = exports.AnnotationOpacity = void 0;
exports.isEventAnnotationLayer = isEventAnnotationLayer;
exports.isFormulaAnnotationLayer = isFormulaAnnotationLayer;
exports.isIntervalAnnotationLayer = isIntervalAnnotationLayer;
exports.isRecordAnnotationResult = isRecordAnnotationResult;
exports.isTableAnnotationLayer = isTableAnnotationLayer;
exports.isTimeseriesAnnotationLayer = isTimeseriesAnnotationLayer;
exports.isTimeseriesAnnotationResult = isTimeseriesAnnotationResult;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint camelcase: 0 */
let AnnotationType;
exports.AnnotationType = AnnotationType;

(function (AnnotationType) {
  AnnotationType["Event"] = "EVENT";
  AnnotationType["Formula"] = "FORMULA";
  AnnotationType["Interval"] = "INTERVAL";
  AnnotationType["Timeseries"] = "TIME_SERIES";
})(AnnotationType || (exports.AnnotationType = AnnotationType = {}));

let AnnotationSourceType;
exports.AnnotationSourceType = AnnotationSourceType;

(function (AnnotationSourceType) {
  AnnotationSourceType["Line"] = "line";
  AnnotationSourceType["Native"] = "NATIVE";
  AnnotationSourceType["Table"] = "table";
  AnnotationSourceType["Undefined"] = "";
})(AnnotationSourceType || (exports.AnnotationSourceType = AnnotationSourceType = {}));

let AnnotationOpacity;
exports.AnnotationOpacity = AnnotationOpacity;

(function (AnnotationOpacity) {
  AnnotationOpacity["High"] = "opacityHigh";
  AnnotationOpacity["Low"] = "opacityLow";
  AnnotationOpacity["Medium"] = "opacityMedium";
  AnnotationOpacity["Undefined"] = "";
})(AnnotationOpacity || (exports.AnnotationOpacity = AnnotationOpacity = {}));

let AnnotationStyle;
exports.AnnotationStyle = AnnotationStyle;

(function (AnnotationStyle) {
  AnnotationStyle["Dashed"] = "dashed";
  AnnotationStyle["Dotted"] = "dotted";
  AnnotationStyle["Solid"] = "solid";
  AnnotationStyle["LongDashed"] = "longDashed";
})(AnnotationStyle || (exports.AnnotationStyle = AnnotationStyle = {}));

function isFormulaAnnotationLayer(layer) {
  return layer.annotationType === AnnotationType.Formula;
}

function isEventAnnotationLayer(layer) {
  return layer.annotationType === AnnotationType.Event;
}

function isIntervalAnnotationLayer(layer) {
  return layer.annotationType === AnnotationType.Interval;
}

function isTimeseriesAnnotationLayer(layer) {
  return layer.annotationType === AnnotationType.Timeseries;
}

function isTableAnnotationLayer(layer) {
  return layer.sourceType === AnnotationSourceType.Table;
}

function isTimeseriesAnnotationResult(result) {
  return Array.isArray(result);
}

function isRecordAnnotationResult(result) {
  return 'columns' in result && 'records' in result;
}
