// @superset-ui/core/lib/query/types/PostProcessing.js
"use strict";

exports.__esModule = true;
exports.PandasAxis = void 0;
exports.isPostProcessingAggregation = isPostProcessingAggregation;
exports.isPostProcessingBoxplot = isPostProcessingBoxplot;
exports.isPostProcessingCompare = isPostProcessingCompare;
exports.isPostProcessingContribution = isPostProcessingContribution;
exports.isPostProcessingCum = isPostProcessingCum;
exports.isPostProcessingDiff = isPostProcessingDiff;
exports.isPostProcessingPivot = isPostProcessingPivot;
exports.isPostProcessingProphet = isPostProcessingProphet;
exports.isPostProcessingResample = isPostProcessingResample;
exports.isPostProcessingRolling = isPostProcessingRolling;
exports.isPostProcessingSort = isPostProcessingSort;

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
let PandasAxis;
exports.PandasAxis = PandasAxis;

(function (PandasAxis) {
  PandasAxis[PandasAxis["Row"] = 0] = "Row";
  PandasAxis[PandasAxis["Column"] = 1] = "Column";
})(PandasAxis || (exports.PandasAxis = PandasAxis = {}));

function isPostProcessingAggregation(rule) {
  return (rule == null ? void 0 : rule.operation) === 'aggregation';
}

function isPostProcessingBoxplot(rule) {
  return (rule == null ? void 0 : rule.operation) === 'boxplot';
}

function isPostProcessingContribution(rule) {
  return (rule == null ? void 0 : rule.operation) === 'contribution';
}

function isPostProcessingPivot(rule) {
  return (rule == null ? void 0 : rule.operation) === 'pivot';
}

function isPostProcessingProphet(rule) {
  return (rule == null ? void 0 : rule.operation) === 'prophet';
}

function isPostProcessingDiff(rule) {
  return (rule == null ? void 0 : rule.operation) === 'diff';
}

function isPostProcessingRolling(rule) {
  return (rule == null ? void 0 : rule.operation) === 'rolling';
}

function isPostProcessingCum(rule) {
  return (rule == null ? void 0 : rule.operation) === 'cum';
}

function isPostProcessingCompare(rule) {
  return (rule == null ? void 0 : rule.operation) === 'compare';
}

function isPostProcessingSort(rule) {
  return (rule == null ? void 0 : rule.operation) === 'sort';
}

function isPostProcessingResample(rule) {
  return (rule == null ? void 0 : rule.operation) === 'resample';
}
