// @encodable/color/lib/scheme/presets/d3Schemes.js
"use strict";

exports.__esModule = true;
exports.d3CyclicalSchemes = exports.d3MultiHueSchemes = exports.d3SingleHueSchemes = exports.d3SequentialSchemes = exports.d3DivergingSchemes = exports.d3CategoricalSchemes = exports.d3Schemes = void 0;

var _d3ScaleChromatic = require("d3-scale-chromatic");

const d3CategoricalSchemes = [{
  type: 'categorical',
  id: 'd3.category10',
  colors: _d3ScaleChromatic.schemeCategory10,
  label: "D3: Category10",
  description: 'Classic Ten categorical colors'
}, {
  type: 'categorical',
  id: 'd3.accent',
  colors: _d3ScaleChromatic.schemeAccent,
  label: "D3: Accent"
}, {
  type: 'categorical',
  id: 'd3.dark2',
  colors: _d3ScaleChromatic.schemeDark2,
  label: "D3: Dark2"
}, {
  type: 'categorical',
  id: 'd3.paired',
  colors: _d3ScaleChromatic.schemePaired,
  label: "D3: Paired"
}, {
  type: 'categorical',
  id: 'd3.pastel1',
  colors: _d3ScaleChromatic.schemePastel1,
  label: "D3: Pastel1"
}, {
  type: 'categorical',
  id: 'd3.pastel2',
  colors: _d3ScaleChromatic.schemePastel2,
  label: "D3: Pastel2"
}, {
  type: 'categorical',
  id: 'd3.set1',
  colors: _d3ScaleChromatic.schemeSet1,
  label: "D3: Set1"
}, {
  type: 'categorical',
  id: 'd3.set2',
  colors: _d3ScaleChromatic.schemeSet2,
  label: "D3: Set2"
}, {
  type: 'categorical',
  id: 'd3.set3',
  colors: _d3ScaleChromatic.schemeSet3,
  label: "D3: Set3"
}, {
  type: 'categorical',
  id: 'd3.tableau10',
  colors: _d3ScaleChromatic.schemeTableau10,
  label: "D3: Tableau10",
  description: 'Ten categorical colors authored by Tableau as part of Tableau 10'
}];
exports.d3CategoricalSchemes = d3CategoricalSchemes;
const d3DivergingSchemes = [{
  type: 'diverging',
  id: 'd3.RdBu',
  interpolator: _d3ScaleChromatic.interpolateRdBu,
  colors: _d3ScaleChromatic.schemeRdBu,
  label: 'D3: Red/Blue'
}, {
  type: 'diverging',
  id: 'd3.BrBG',
  interpolator: _d3ScaleChromatic.interpolateBrBG,
  colors: _d3ScaleChromatic.schemeBrBG,
  label: 'D3: Brown/BlueGreen'
}, {
  type: 'diverging',
  id: 'd3.PRGn',
  interpolator: _d3ScaleChromatic.interpolatePRGn,
  colors: _d3ScaleChromatic.schemePRGn,
  label: 'D3: PurpleRed/Green'
}, {
  type: 'diverging',
  id: 'd3.PiYG',
  interpolator: _d3ScaleChromatic.interpolatePiYG,
  colors: _d3ScaleChromatic.schemePiYG,
  label: 'D3: Pink/YellowGreen'
}, {
  type: 'diverging',
  id: 'd3.PuOr',
  interpolator: _d3ScaleChromatic.interpolatePuOr,
  colors: _d3ScaleChromatic.schemePuOr,
  label: 'D3: Purple/Orange'
}, {
  type: 'diverging',
  id: 'd3.RdGy',
  interpolator: _d3ScaleChromatic.interpolateRdGy,
  colors: _d3ScaleChromatic.schemeRdGy,
  label: 'D3: Red/Grey'
}, {
  type: 'diverging',
  id: 'd3.RdYlBu',
  interpolator: _d3ScaleChromatic.interpolateRdYlBu,
  colors: _d3ScaleChromatic.schemeRdYlBu,
  label: 'D3: Red/Yellow/Blue'
}, {
  type: 'diverging',
  id: 'd3.RdYlGn',
  interpolator: _d3ScaleChromatic.interpolateRdYlGn,
  colors: _d3ScaleChromatic.schemeRdYlGn,
  label: 'D3: Red/Yellow/Green'
}, {
  type: 'diverging',
  id: 'd3.spectral',
  interpolator: _d3ScaleChromatic.interpolateSpectral,
  colors: _d3ScaleChromatic.schemeSpectral,
  label: 'D3: Spectral'
}];
exports.d3DivergingSchemes = d3DivergingSchemes;
const d3SingleHueSchemes = [{
  type: 'sequential',
  id: 'd3.blues',
  interpolator: _d3ScaleChromatic.interpolateBlues,
  colors: _d3ScaleChromatic.schemeBlues,
  label: 'D3: Blues'
}, {
  type: 'sequential',
  id: 'd3.greens',
  interpolator: _d3ScaleChromatic.interpolateGreens,
  colors: _d3ScaleChromatic.schemeGreens,
  label: 'D3: Greens'
}, {
  type: 'sequential',
  id: 'd3.greys',
  interpolator: _d3ScaleChromatic.interpolateGreys,
  colors: _d3ScaleChromatic.schemeGreys,
  label: 'D3: Greys'
}, {
  type: 'sequential',
  id: 'd3.oranges',
  interpolator: _d3ScaleChromatic.interpolateOranges,
  colors: _d3ScaleChromatic.schemeOranges,
  label: 'D3: Oranges'
}, {
  type: 'sequential',
  id: 'd3.purples',
  interpolator: _d3ScaleChromatic.interpolatePurples,
  colors: _d3ScaleChromatic.schemePurples,
  label: 'D3: Purples'
}, {
  type: 'sequential',
  id: 'd3.reds',
  interpolator: _d3ScaleChromatic.interpolateReds,
  colors: _d3ScaleChromatic.schemeReds,
  label: 'D3: Reds'
}];
exports.d3SingleHueSchemes = d3SingleHueSchemes;
const d3MultiHueSchemes = [{
  type: 'sequential',
  id: 'd3.turbo',
  interpolator: _d3ScaleChromatic.interpolateTurbo,
  label: 'D3: Turbo'
}, {
  type: 'sequential',
  id: 'd3.viridis',
  interpolator: _d3ScaleChromatic.interpolateViridis,
  label: 'D3: Viridis'
}, {
  type: 'sequential',
  id: 'd3.inferno',
  interpolator: _d3ScaleChromatic.interpolateInferno,
  label: 'D3: Inferno'
}, {
  type: 'sequential',
  id: 'd3.magma',
  interpolator: _d3ScaleChromatic.interpolateMagma,
  label: 'D3: Magma'
}, {
  type: 'sequential',
  id: 'd3.plasma',
  interpolator: _d3ScaleChromatic.interpolatePlasma,
  label: 'D3: Plasma'
}, {
  type: 'sequential',
  id: 'd3.cividis',
  interpolator: _d3ScaleChromatic.interpolateCividis,
  label: 'D3: Cividis'
}, {
  type: 'sequential',
  id: 'd3.warm',
  interpolator: _d3ScaleChromatic.interpolateWarm,
  label: 'D3: Warm'
}, {
  type: 'sequential',
  id: 'd3.cool',
  interpolator: _d3ScaleChromatic.interpolateCool,
  label: 'D3: Cool'
}, {
  type: 'sequential',
  id: 'd3.cubehelixDefault',
  interpolator: _d3ScaleChromatic.interpolateCubehelixDefault,
  label: 'D3: Cube Helix'
}, {
  type: 'sequential',
  id: 'd3.BuGn',
  interpolator: _d3ScaleChromatic.interpolateBuGn,
  colors: _d3ScaleChromatic.schemeBuGn,
  label: 'D3: Blue/Green'
}, {
  type: 'sequential',
  id: 'd3.BuPu',
  interpolator: _d3ScaleChromatic.interpolateBuPu,
  colors: _d3ScaleChromatic.schemeBuPu,
  label: 'D3: Blue/Purple'
}, {
  type: 'sequential',
  id: 'd3.GnBu',
  interpolator: _d3ScaleChromatic.interpolateGnBu,
  colors: _d3ScaleChromatic.schemeGnBu,
  label: 'D3: Green/Blue'
}, {
  type: 'sequential',
  id: 'd3.OrRd',
  interpolator: _d3ScaleChromatic.interpolateOrRd,
  colors: _d3ScaleChromatic.schemeOrRd,
  label: 'D3: Orange/Red'
}, {
  type: 'sequential',
  id: 'd3.PuBuGn',
  interpolator: _d3ScaleChromatic.interpolatePuBuGn,
  colors: _d3ScaleChromatic.schemePuBuGn,
  label: 'D3: Purple/Blue/Green'
}, {
  type: 'sequential',
  id: 'd3.PuBu',
  interpolator: _d3ScaleChromatic.interpolatePuBu,
  colors: _d3ScaleChromatic.schemePuBu,
  label: 'D3: Purple/Blue'
}, {
  type: 'sequential',
  id: 'd3.PuRd',
  interpolator: _d3ScaleChromatic.interpolatePuRd,
  colors: _d3ScaleChromatic.schemePuRd,
  label: 'D3: Purple/Red'
}, {
  type: 'sequential',
  id: 'd3.RdPu',
  interpolator: _d3ScaleChromatic.interpolateRdPu,
  colors: _d3ScaleChromatic.schemeRdPu,
  label: 'D3: Red/Purple'
}, {
  type: 'sequential',
  id: 'd3.YlGnBu',
  interpolator: _d3ScaleChromatic.interpolateYlGnBu,
  colors: _d3ScaleChromatic.schemeYlGnBu,
  label: 'D3: Yellow/Green/Blue'
}, {
  type: 'sequential',
  id: 'd3.YlGn',
  interpolator: _d3ScaleChromatic.interpolateYlGn,
  colors: _d3ScaleChromatic.schemeYlGn,
  label: 'D3: Yellow/Green'
}, {
  type: 'sequential',
  id: 'd3.YlOrBr',
  interpolator: _d3ScaleChromatic.interpolateYlOrBr,
  colors: _d3ScaleChromatic.schemeYlOrBr,
  label: 'D3: Yellow/Orange/Brown'
}, {
  type: 'sequential',
  id: 'd3.YlOrRd',
  interpolator: _d3ScaleChromatic.interpolateYlOrRd,
  colors: _d3ScaleChromatic.schemeYlOrRd,
  label: 'D3: Yellow/Orange/Red'
}];
exports.d3MultiHueSchemes = d3MultiHueSchemes;
const d3CyclicalSchemes = [{
  type: 'sequential',
  id: 'd3.rainbow',
  interpolator: _d3ScaleChromatic.interpolateRainbow,
  label: 'D3: Rainbow'
}, {
  type: 'sequential',
  id: 'd3.sinebow',
  interpolator: _d3ScaleChromatic.interpolateSinebow,
  label: 'D3: Sinebow'
}];
exports.d3CyclicalSchemes = d3CyclicalSchemes;
const d3SequentialSchemes = d3SingleHueSchemes.concat(d3MultiHueSchemes).concat(d3CyclicalSchemes);
exports.d3SequentialSchemes = d3SequentialSchemes;
const d3Schemes = d3CategoricalSchemes.concat(d3DivergingSchemes).concat(d3SequentialSchemes);
exports.d3Schemes = d3Schemes;
