// @superset-ui/legacy-preset-chart-deckgl/lib/layers/Scatter/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _chartControls = require("@superset-ui/chart-controls");

var _core = require("@superset-ui/core");

var _controls = _interopRequireDefault(require("../../utilities/controls"));

var _Shared_DeckGL = require("../../utilities/Shared_DeckGL");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = {
  onInit: controlState => _extends({}, controlState, {
    time_grain_sqla: _extends({}, controlState.time_grain_sqla, {
      value: null
    }),
    granularity: _extends({}, controlState.granularity, {
      value: null
    })
  }),
  controlPanelSections: [_chartControls.sections.legacyRegularTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [[_Shared_DeckGL.spatial, null], ['row_limit', _Shared_DeckGL.filterNulls], ['adhoc_filters']]
  }, {
    label: (0, _core.t)('Map'),
    expanded: true,
    controlSetRows: [[_Shared_DeckGL.mapboxStyle, _Shared_DeckGL.viewport], [_Shared_DeckGL.autozoom, null]]
  }, {
    label: (0, _core.t)('Point Size'),
    controlSetRows: [[_Shared_DeckGL.pointRadiusFixed], [{
      name: 'point_unit',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Point Unit'),
        default: 'square_m',
        clearable: false,
        choices: [['square_m', 'Square meters'], ['square_km', 'Square kilometers'], ['square_miles', 'Square miles'], ['radius_m', 'Radius in meters'], ['radius_km', 'Radius in kilometers'], ['radius_miles', 'Radius in miles']],
        description: (0, _core.t)('The unit of measure for the specified point radius')
      }
    }], [{
      name: 'min_radius',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Minimum Radius'),
        isFloat: true,
        validators: [_core.validateNonEmpty],
        renderTrigger: true,
        default: 2,
        description: (0, _core.t)('Minimum radius size of the circle, in pixels. As the zoom level changes, this ' + 'insures that the circle respects this minimum radius.')
      }
    }, {
      name: 'max_radius',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Maximum Radius'),
        isFloat: true,
        validators: [_core.validateNonEmpty],
        renderTrigger: true,
        default: 250,
        description: (0, _core.t)('Maxium radius size of the circle, in pixels. As the zoom level changes, this ' + 'insures that the circle respects this maximum radius.')
      }
    }], [_Shared_DeckGL.multiplier, null]]
  }, {
    label: (0, _core.t)('Point Color'),
    controlSetRows: [['color_picker'], [_Shared_DeckGL.legendPosition], [_Shared_DeckGL.legendFormat], [_extends({}, _Shared_DeckGL.dimension, {
      label: (0, _core.t)('Categorical Color'),
      description: (0, _core.t)('Pick a dimension from which categorical colors are defined')
    })], ['color_scheme']]
  }, {
    label: (0, _core.t)('Advanced'),
    controlSetRows: [[_Shared_DeckGL.jsColumns], [_Shared_DeckGL.jsDataMutator], [_Shared_DeckGL.jsTooltip], [_Shared_DeckGL.jsOnclickHref]]
  }],
  controlOverrides: {
    size: {
      validators: []
    },
    time_grain_sqla: _controls.default
  }
};
exports.default = _default;
