// @applitools/core/dist/classic/check-and-close.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCheckAndClose = void 0;
const driver_1 = require("@applitools/driver");
const take_screenshot_1 = require("../automation/utils/take-screenshot");
const take_dom_capture_1 = require("./utils/take-dom-capture");
const to_base_check_settings_1 = require("../utils/to-base-check-settings");
const wait_for_lazy_load_1 = require("../utils/wait-for-lazy-load");
const utils = __importStar(require("@applitools/utils"));
function makeCheckAndClose({ eyes, target: defaultTarget, spec, logger: defaultLogger, }) {
    return async function checkAndClose({ target = defaultTarget, settings = {}, logger = defaultLogger, } = {}) {
        var _a;
        logger.log('Command "checkAndClose" is called with settings', settings);
        const baseEyes = await eyes.getBaseEyes({ logger });
        if (!(0, driver_1.isDriver)(target, spec)) {
            const baseSettings = settings;
            return (await Promise.all(baseEyes.map(baseEyes => baseEyes.checkAndClose({ target, settings: baseSettings, logger })))).flat();
        }
        const driver = await (0, driver_1.makeDriver)({ spec, driver: target, logger });
        if (settings.lazyLoad && driver.isWeb) {
            await (0, wait_for_lazy_load_1.waitForLazyLoad)({
                context: driver.currentContext,
                settings: settings.lazyLoad !== true ? settings.lazyLoad : {},
                logger,
            });
        }
        const { elementReferencesToCalculate, getBaseCheckSettings } = (0, to_base_check_settings_1.toBaseCheckSettings)({ settings });
        const screenshot = await (0, take_screenshot_1.takeScreenshot)({
            driver,
            settings: { ...settings, regionsToCalculate: elementReferencesToCalculate },
            logger,
        });
        const baseTarget = {
            name: await driver.getTitle(),
            source: await driver.getUrl(),
            image: await screenshot.image.toPng(),
            locationInViewport: utils.geometry.location(screenshot.region),
            isTransformed: true,
        };
        const baseSettings = getBaseCheckSettings({ calculatedRegions: screenshot.calculatedRegions });
        if (driver.isWeb && settings.sendDom) {
            if (settings.fully)
                await screenshot.scrollingElement.setAttribute('data-applitools-scroll', 'true');
            else
                await ((_a = screenshot.element) === null || _a === void 0 ? void 0 : _a.setAttribute('data-applitools-scroll', 'true'));
            baseTarget.dom = await (0, take_dom_capture_1.takeDomCapture)({ driver, logger }).catch(() => null);
        }
        if (settings.pageId) {
            const scrollingElement = await driver.mainContext.getScrollingElement();
            const scrollingOffset = driver.isNative ? { x: 0, y: 0 } : await scrollingElement.getScrollOffset();
            baseTarget.locationInView = utils.geometry.offset(scrollingOffset, screenshot.region);
            baseTarget.fullViewSize = scrollingElement ? await scrollingElement.getContentSize() : await driver.getViewportSize();
        }
        await screenshot.restoreState();
        return (await Promise.all(baseEyes.map(baseEyes => baseEyes.checkAndClose({ target: baseTarget, settings: baseSettings, logger })))).flat();
    };
}
exports.makeCheckAndClose = makeCheckAndClose;
