// @superset-ui/core/lib/color/colorSchemes/categorical/index.js
"use strict";

exports.__esModule = true;
exports.CategoricalSuperset = exports.CategoricalPreset = exports.CategoricalLyft = exports.CategoricalGoogle = exports.CategoricalEcharts = exports.CategoricalD3 = exports.CategoricalAirbnb = void 0;

var _airbnb = _interopRequireDefault(require("./airbnb"));

exports.CategoricalAirbnb = _airbnb.default;

var _d = _interopRequireDefault(require("./d3"));

exports.CategoricalD3 = _d.default;

var _echarts = _interopRequireDefault(require("./echarts"));

exports.CategoricalEcharts = _echarts.default;

var _google = _interopRequireDefault(require("./google"));

exports.CategoricalGoogle = _google.default;

var _lyft = _interopRequireDefault(require("./lyft"));

exports.CategoricalLyft = _lyft.default;

var _preset = _interopRequireDefault(require("./preset"));

exports.CategoricalPreset = _preset.default;

var _superset = _interopRequireDefault(require("./superset"));

exports.CategoricalSuperset = _superset.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
