// @superset-ui/core/lib/query/api/v1/types.js
"use strict";

exports.__esModule = true;
exports.SupersetApiErrorType = exports.SupersetApiError = void 0;

/* eslint-disable camelcase */

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * Runtime options when calling a Superset API. Currently only allow overriding
 * SupersetClient instance.
 */

/**
 * Superset API error types.
 * Ref: https://github.com/apache/incubator-superset/blob/318e5347bc6f88119725775baa4ab9a398a6f0b0/superset/errors.py#L24
 *
 * TODO: migrate superset-frontend/src/components/ErrorMessage/types.ts over
 */
let SupersetApiErrorType;
/**
 * API Error json response from the backend (or fetch API in the frontend).
 * See SIP-40 and SIP-41: https://github.com/apache/incubator-superset/issues/9298
 */

exports.SupersetApiErrorType = SupersetApiErrorType;

(function (SupersetApiErrorType) {
  SupersetApiErrorType["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
  SupersetApiErrorType["FRONTEND_CSRF_ERROR"] = "FRONTEND_CSRF_ERROR";
  SupersetApiErrorType["FRONTEND_NETWORK_ERROR"] = "FRONTEND_NETWORK_ERROR";
  SupersetApiErrorType["FRONTEND_TIMEOUT_ERROR"] = "FRONTEND_TIMEOUT_ERROR";
  SupersetApiErrorType["GENERIC_DB_ENGINE_ERROR"] = "GENERIC_DB_ENGINE_ERROR";
  SupersetApiErrorType["VIZ_GET_DF_ERROR"] = "VIZ_GET_DF_ERROR";
  SupersetApiErrorType["UNKNOWN_DATASOURCE_TYPE_ERROR"] = "UNKNOWN_DATASOURCE_TYPE_ERROR";
  SupersetApiErrorType["FAILED_FETCHING_DATASOURCE_INFO_ERROR"] = "FAILED_FETCHING_DATASOURCE_INFO_ERROR";
  SupersetApiErrorType["TABLE_SECURITY_ACCESS_ERROR"] = "TABLE_SECURITY_ACCESS_ERROR";
  SupersetApiErrorType["DATASOURCE_SECURITY_ACCESS_ERROR"] = "DATASOURCE_SECURITY_ACCESS_ERROR";
  SupersetApiErrorType["MISSING_OWNERSHIP_ERROR"] = "MISSING_OWNERSHIP_ERROR";
})(SupersetApiErrorType || (exports.SupersetApiErrorType = SupersetApiErrorType = {}));

class SupersetApiError extends Error {
  constructor({
    status,
    statusText,
    message,
    link,
    extra,
    stack,
    error_type: errorType,
    originalError
  }) {
    super(message);
    this.status = void 0;
    this.statusText = void 0;
    this.errorType = void 0;
    this.extra = void 0;
    this.originalError = void 0;
    const originalErrorStack = stack || (originalError instanceof Error ? originalError.stack : undefined);
    this.stack = originalErrorStack && this.stack ? [this.stack.split('\n')[0], ...originalErrorStack.split('\n').slice(1)].join('\n') : this.stack;
    this.name = 'SupersetApiError';
    this.errorType = errorType || SupersetApiErrorType.UNKNOWN_ERROR;
    this.extra = extra || {};

    if (link) {
      this.extra.link = link;
    }

    this.status = status;
    this.statusText = statusText;
    this.originalError = originalError;
  }

}

exports.SupersetApiError = SupersetApiError;
