// @superset-ui/legacy-plugin-chart-time-table/lib/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const config = {
  controlPanelSections: [_chartControls.sections.legacyTimeseriesTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [['metrics'], ['adhoc_filters'], ['groupby'], ['limit'], [{
      name: 'column_collection',
      config: {
        type: 'CollectionControl',
        label: (0, _core.t)('Time Series Columns'),
        validators: [_core.validateNonEmpty],
        controlName: 'TimeSeriesColumnControl'
      }
    }], [{
      name: 'url',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('URL'),
        description: (0, _core.t)("Templated link, it's possible to include {{ metric }} " + 'or other values coming from the controls.'),
        default: ''
      }
    }]]
  }],
  controlOverrides: {
    groupby: {
      multiple: false
    }
  }
};
var _default = config;
exports.default = _default;
