// @encodable/format/lib/number/index.js
"use strict";

exports.__esModule = true;
exports.getNumberFormatter = getNumberFormatter;
exports.formatNumber = formatNumber;
exports.createSmartNumberFormatter = exports.createD3NumberFormatter = exports.previewNumber = exports.createNumberFormatter = exports.NumberFormats = exports.getNumberFormatterRegistry = void 0;

var _registry = require("@encodable/registry");

var _NumberFormatterRegistry = _interopRequireDefault(require("./NumberFormatterRegistry"));

exports.NumberFormatterRegistry = _NumberFormatterRegistry.default;

var _NumberFormats = _interopRequireDefault(require("./NumberFormats"));

exports.NumberFormats = _NumberFormats.default;

var _createNumberFormatter = _interopRequireDefault(require("./createNumberFormatter"));

exports.createNumberFormatter = _createNumberFormatter.default;

var _previewNumber = _interopRequireDefault(require("./previewNumber"));

exports.previewNumber = _previewNumber.default;

var _createD3NumberFormatter = _interopRequireDefault(require("./factories/createD3NumberFormatter"));

exports.createD3NumberFormatter = _createD3NumberFormatter.default;

var _createSmartNumberFormatter = _interopRequireDefault(require("./factories/createSmartNumberFormatter"));

exports.createSmartNumberFormatter = _createSmartNumberFormatter.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getNumberFormatterRegistry = (0, _registry.makeSingleton)(() => new _NumberFormatterRegistry.default({
  globalId: '@encodable/format:NumberFormatterRegistry'
}));
exports.getNumberFormatterRegistry = getNumberFormatterRegistry;

function getNumberFormatter(format) {
  return getNumberFormatterRegistry().get(format);
}

function formatNumber(format, value) {
  return getNumberFormatterRegistry().format(format, value);
}
