// @superset-ui/legacy-plugin-chart-histogram/lib/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const allColumns = {
  type: 'SelectControl',
  label: (0, _core.t)('Columns'),
  default: null,
  description: (0, _core.t)('Select the numeric columns to draw the histogram'),
  mapStateToProps: state => ({
    choices: (0, _chartControls.columnChoices)(state.datasource)
  }),
  multi: true,
  validators: [_core.validateNonEmpty]
};
const dndAllColumns = { ..._chartControls.dndColumnsControl,
  description: (0, _core.t)('Select the numeric columns to draw the histogram'),
  validators: [_core.validateNonEmpty]
};
const columnsConfig = (0, _core.isFeatureEnabled)(_core.FeatureFlag.ENABLE_EXPLORE_DRAG_AND_DROP) ? dndAllColumns : allColumns;
const config = {
  controlPanelSections: [_chartControls.sections.legacyRegularTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [[{
      name: 'all_columns_x',
      config: columnsConfig
    }], ['adhoc_filters'], ['row_limit'], ['groupby']]
  }, {
    label: (0, _core.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_scheme'], [{
      name: 'link_length',
      config: {
        type: 'SelectControl',
        renderTrigger: true,
        freeForm: true,
        label: (0, _core.t)('No of Bins'),
        default: 5,
        choices: (0, _chartControls.formatSelectOptions)(['10', '25', '50', '75', '100', '150', '200', '250']),
        description: (0, _core.t)('Select the number of bins for the histogram')
      }
    }], [{
      name: 'x_axis_label',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('X Axis Label'),
        renderTrigger: true,
        default: ''
      }
    }], [{
      name: 'y_axis_label',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Y Axis Label'),
        renderTrigger: true,
        default: ''
      }
    }], [{
      name: 'show_legend',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Legend'),
        renderTrigger: true,
        default: false,
        description: (0, _core.t)('Whether to display the legend (toggles)')
      }
    }], [{
      name: 'normalized',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Normalized'),
        renderTrigger: true,
        description: (0, _core.t)('Whether to normalize the histogram'),
        default: false
      }
    }], [{
      name: 'cumulative',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Cumulative'),
        renderTrigger: true,
        description: (0, _core.t)('Whether to make the histogram cumulative'),
        default: false
      }
    }]]
  }]
};
var _default = config;
exports.default = _default;
