// @vx/shape/build/shapes/Arc.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Arc;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _d3Shape = require('d3-shape');

var _additionalProps = require('../util/additionalProps');

var _additionalProps2 = _interopRequireDefault(_additionalProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

Arc.propTypes = {
  className: _propTypes2.default.string,
  data: _propTypes2.default.any,
  centroid: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  innerRadius: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  outerRadius: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  cornerRadius: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  startAngle: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  endAngle: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  padAngle: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
  padRadius: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number])
};

function Arc(_ref) {
  var className = _ref.className,
      data = _ref.data,
      centroid = _ref.centroid,
      innerRadius = _ref.innerRadius,
      outerRadius = _ref.outerRadius,
      cornerRadius = _ref.cornerRadius,
      startAngle = _ref.startAngle,
      endAngle = _ref.endAngle,
      padAngle = _ref.padAngle,
      padRadius = _ref.padRadius,
      restProps = _objectWithoutProperties(_ref, ['className', 'data', 'centroid', 'innerRadius', 'outerRadius', 'cornerRadius', 'startAngle', 'endAngle', 'padAngle', 'padRadius']);

  var arc = (0, _d3Shape.arc)();
  if (centroid) arc.centroid(centroid);
  if (innerRadius) arc.innerRadius(innerRadius);
  if (outerRadius) arc.outerRadius(outerRadius);
  if (cornerRadius) arc.cornerRadius(cornerRadius);
  if (startAngle) arc.startAngle(startAngle);
  if (endAngle) arc.endAngle(endAngle);
  if (padAngle) arc.padAngle(padAngle);
  if (padRadius) arc.padRadius(padRadius);
  return _react2.default.createElement('path', _extends({ className: (0, _classnames2.default)('vx-arc', className), d: arc(data) }, (0, _additionalProps2.default)(restProps, data)));
}
