// @superset-ui/legacy-preset-chart-big-number/lib/BigNumber/BigNumber.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
exports.renderTooltipFactory = renderTooltipFactory;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _shortid = _interopRequireDefault(require("shortid"));

var _core = require("@superset-ui/core");

var _xyChart = require("@data-ui/xy-chart");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const defaultNumberFormatter = (0, _core.getNumberFormatter)();
const CHART_MARGIN = {
  top: 4,
  right: 4,
  bottom: 4,
  left: 4
};
const PROPORTION = {
  // text size: proportion of the chart container sans trendline
  KICKER: 0.1,
  HEADER: 0.3,
  SUBHEADER: 0.125,
  // trendline size: proportion of the whole chart container
  TRENDLINE: 0.3
};

function renderTooltipFactory(formatDate = _core.smartDateVerboseFormatter, formatValue = defaultNumberFormatter) {
  return function renderTooltip({
    datum: {
      x,
      y
    }
  }) {
    // even though `formatDate` supports timestamp as numbers, we need
    // `new Date` to pass type check
    return (0, _react2.jsx)("div", {
      style: {
        padding: '4px 8px'
      }
    }, formatDate(new Date(x)), (0, _react2.jsx)("br", null), (0, _react2.jsx)("strong", null, y === null ? (0, _core.t)('N/A') : formatValue(y)));
  };
}

class BigNumberVis extends _react.default.PureComponent {
  constructor(...args) {
    super(...args);
    this.gradientId = _shortid.default.generate();
  }

  getClassName() {
    const {
      className,
      showTrendLine,
      bigNumberFallback
    } = this.props;
    const names = `superset-legacy-chart-big-number ${className} ${bigNumberFallback ? 'is-fallback-value' : ''}`;
    if (showTrendLine) return names;
    return `${names} no-trendline`;
  }

  createTemporaryContainer() {
    const container = document.createElement('div');
    container.className = this.getClassName();
    container.style.position = 'absolute'; // so it won't disrupt page layout

    container.style.opacity = '0'; // and not visible

    return container;
  }

  renderFallbackWarning() {
    const {
      bigNumberFallback,
      formatTime,
      showTimestamp
    } = this.props;
    if (!bigNumberFallback || showTimestamp) return null;
    return (0, _react2.jsx)("span", {
      className: "alert alert-warning",
      role: "alert",
      title: (0, _core.t)(`Last available value seen on %s`, formatTime(bigNumberFallback.x))
    }, (0, _core.t)('Not up to date'));
  }

  renderKicker(maxHeight) {
    const {
      timestamp,
      showTimestamp,
      formatTime,
      width
    } = this.props;
    if (!showTimestamp) return null;
    const text = timestamp === null ? '' : formatTime(timestamp);
    const container = this.createTemporaryContainer();
    document.body.append(container);
    const fontSize = (0, _core.computeMaxFontSize)({
      text,
      maxWidth: width,
      maxHeight,
      className: 'kicker',
      container
    });
    container.remove();
    return (0, _react2.jsx)("div", {
      className: "kicker",
      style: {
        fontSize,
        height: maxHeight
      }
    }, text);
  }

  renderHeader(maxHeight) {
    const {
      bigNumber,
      headerFormatter,
      width
    } = this.props;
    const text = bigNumber === null ? (0, _core.t)('No data') : headerFormatter(bigNumber);
    const container = this.createTemporaryContainer();
    document.body.append(container);
    const fontSize = (0, _core.computeMaxFontSize)({
      text,
      maxWidth: width,
      maxHeight,
      className: 'header-line',
      container
    });
    container.remove();
    return (0, _react2.jsx)("div", {
      className: "header-line",
      style: {
        fontSize,
        height: maxHeight
      }
    }, text);
  }

  renderSubheader(maxHeight) {
    const {
      bigNumber,
      subheader,
      width,
      bigNumberFallback
    } = this.props;
    let fontSize = 0;
    const NO_DATA_OR_HASNT_LANDED = (0, _core.t)('No data after filtering or data is NULL for the latest time record');
    const NO_DATA = (0, _core.t)('Try applying different filters or ensuring your datasource has data');
    let text = subheader;

    if (bigNumber === null) {
      text = bigNumberFallback ? NO_DATA : NO_DATA_OR_HASNT_LANDED;
    }

    if (text) {
      const container = this.createTemporaryContainer();
      document.body.append(container);
      fontSize = (0, _core.computeMaxFontSize)({
        text,
        maxWidth: width,
        maxHeight,
        className: 'subheader-line',
        container
      });
      container.remove();
      return (0, _react2.jsx)("div", {
        className: "subheader-line",
        style: {
          fontSize,
          height: maxHeight
        }
      }, text);
    }

    return null;
  }

  renderTrendline(maxHeight) {
    const {
      width,
      trendLineData,
      mainColor,
      subheader,
      startYAxisAtZero,
      headerFormatter,
      formatTime,
      fromDatetime,
      timeRangeFixed
    } = this.props; // if can't find any non-null values, no point rendering the trendline

    if (!(trendLineData != null && trendLineData.some(d => d.y !== null))) {
      return null;
    } // Apply a fixed X range if a time range is specified.
    //
    // XYChart checks the existence of `domain` property and decide whether to
    // apply a domain or not, so it must not be `null` or `undefined`


    const xScale = {
      type: 'timeUtc'
    };
    const tooltipData = trendLineData && [...trendLineData];

    if (tooltipData && timeRangeFixed && fromDatetime) {
      const toDatetime = this.props.toDatetime ?? Date.now();

      if (tooltipData[0].x > fromDatetime) {
        tooltipData.unshift({
          x: fromDatetime,
          y: null
        });
      }

      if (tooltipData[tooltipData.length - 1].x < toDatetime) {
        tooltipData.push({
          x: toDatetime,
          y: null
        });
      }

      xScale.domain = [fromDatetime, toDatetime];
    }

    return (0, _react2.jsx)(_xyChart.XYChart, {
      snapTooltipToDataX: true,
      ariaLabel: `Big number visualization ${subheader}` // headerFormatter always NumberFormatter in BigNumber with treadline
      ,
      renderTooltip: renderTooltipFactory(formatTime, headerFormatter),
      xScale: xScale,
      yScale: {
        type: 'linear',
        includeZero: startYAxisAtZero
      },
      width: Math.floor(width),
      height: maxHeight,
      margin: CHART_MARGIN,
      eventTrigger: "container"
    }, (0, _react2.jsx)(_xyChart.LinearGradient, {
      id: this.gradientId,
      from: mainColor,
      to: "#fff"
    }), (0, _react2.jsx)(_xyChart.AreaSeries, {
      data: tooltipData,
      fill: `url(#${this.gradientId})`,
      stroke: mainColor
    }), (0, _react2.jsx)(_xyChart.CrossHair, {
      fullHeight: true,
      stroke: mainColor,
      circleFill: mainColor,
      circleStroke: "#fff",
      showHorizontalLine: false,
      strokeDasharray: "5,2"
    }));
  }

  render() {
    const {
      showTrendLine,
      height,
      kickerFontSize,
      headerFontSize,
      subheaderFontSize
    } = this.props;
    const className = this.getClassName();

    if (showTrendLine) {
      const chartHeight = Math.floor(PROPORTION.TRENDLINE * height);
      const allTextHeight = height - chartHeight;
      return (0, _react2.jsx)("div", {
        className: className
      }, (0, _react2.jsx)("div", {
        className: "text-container",
        style: {
          height: allTextHeight
        }
      }, this.renderFallbackWarning(), this.renderKicker(Math.ceil(kickerFontSize * (1 - PROPORTION.TRENDLINE) * height)), this.renderHeader(Math.ceil(headerFontSize * (1 - PROPORTION.TRENDLINE) * height)), this.renderSubheader(Math.ceil(subheaderFontSize * (1 - PROPORTION.TRENDLINE) * height))), this.renderTrendline(chartHeight));
    }

    return (0, _react2.jsx)("div", {
      className: className,
      style: {
        height
      }
    }, this.renderFallbackWarning(), this.renderKicker(kickerFontSize * height), this.renderHeader(Math.ceil(headerFontSize * height)), this.renderSubheader(Math.ceil(subheaderFontSize * height)));
  }

}

BigNumberVis.propTypes = {
  className: _propTypes.default.string,
  width: _propTypes.default.number.isRequired,
  height: _propTypes.default.number.isRequired,
  bigNumber: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.oneOf([null])]),
  bigNumberFallback: _propTypes.default.shape({
    x: _propTypes.default.number.isRequired,
    y: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.oneOf([null])])
  }),
  fromDatetime: _propTypes.default.number,
  toDatetime: _propTypes.default.number,
  headerFontSize: _propTypes.default.number,
  kickerFontSize: _propTypes.default.number,
  subheader: _propTypes.default.string,
  subheaderFontSize: _propTypes.default.number,
  showTimestamp: _propTypes.default.bool,
  showTrendLine: _propTypes.default.bool,
  startYAxisAtZero: _propTypes.default.bool,
  timeRangeFixed: _propTypes.default.bool,
  timestamp: _propTypes.default.number,
  trendLineData: _propTypes.default.arrayOf(_propTypes.default.shape({
    x: _propTypes.default.number.isRequired,
    y: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.oneOf([null])])
  })),
  mainColor: _propTypes.default.string
};
BigNumberVis.defaultProps = {
  className: '',
  headerFormatter: defaultNumberFormatter,
  formatTime: _core.smartDateVerboseFormatter,
  headerFontSize: PROPORTION.HEADER,
  kickerFontSize: PROPORTION.KICKER,
  mainColor: _core.BRAND_COLOR,
  showTimestamp: false,
  showTrendLine: false,
  startYAxisAtZero: true,
  subheader: '',
  subheaderFontSize: PROPORTION.SUBHEADER,
  timeRangeFixed: false
};

var _default = (0, _core.styled)(BigNumberVis)`
  font-family: ${({
  theme
}) => theme.typography.families.sansSerif};
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;

  &.no-trendline .subheader-line {
    padding-bottom: 0.3em;
  }

  .text-container {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;
    .alert {
      font-size: ${({
  theme
}) => theme.typography.sizes.s};
      margin: -0.5em 0 0.4em;
      line-height: 1;
      padding: 2px 4px 3px;
      border-radius: 3px;
    }
  }

  .kicker {
    font-weight: ${({
  theme
}) => theme.typography.weights.light};
    line-height: 1em;
    padding-bottom: 2em;
  }

  .header-line {
    font-weight: ${({
  theme
}) => theme.typography.weights.normal};
    position: relative;
    line-height: 1em;
    span {
      position: absolute;
      bottom: 0;
    }
  }

  .subheader-line {
    font-weight: ${({
  theme
}) => theme.typography.weights.light};
    line-height: 1em;
    padding-bottom: 0;
  }

  &.is-fallback-value {
    .kicker,
    .header-line,
    .subheader-line {
      opacity: 0.5;
    }
  }

  .superset-data-ui-tooltip {
    z-index: 1000;
    background: #000;
  }
`;

exports.default = _default;
