// @superset-ui/legacy-preset-chart-big-number/lib/BigNumber/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

var _react = _interopRequireDefault(require("react"));

var _sharedControls = require("../sharedControls");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const config = {
  controlPanelSections: [_chartControls.sections.legacyTimeseriesTime, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [['metric'], ['adhoc_filters']]
  }, {
    label: (0, _core.t)('Options'),
    tabOverride: 'data',
    expanded: true,
    controlSetRows: [[{
      name: 'compare_lag',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Comparison Period Lag'),
        isInt: true,
        description: (0, _core.t)('Based on granularity, number of time periods to compare against')
      }
    }], [{
      name: 'compare_suffix',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Comparison suffix'),
        description: (0, _core.t)('Suffix to apply after the percentage display')
      }
    }], ['y_axis_format'], [{
      name: 'time_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _core.t)('Timestamp format'),
        renderTrigger: true,
        choices: _chartControls.D3_TIME_FORMAT_OPTIONS,
        description: _chartControls.D3_FORMAT_DOCS
      }
    }], [{
      name: 'show_timestamp',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Show Timestamp'),
        renderTrigger: true,
        default: false,
        description: (0, _core.t)('Whether to display the timestamp')
      }
    }], [{
      name: 'show_trend_line',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Show Trend Line'),
        renderTrigger: true,
        default: true,
        description: (0, _core.t)('Whether to display the trend line')
      }
    }], [{
      name: 'start_y_axis_at_zero',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Start y-axis at 0'),
        renderTrigger: true,
        default: true,
        description: (0, _core.t)('Start y-axis at zero. Uncheck to start y-axis at minimum value in the data.')
      }
    }], [{
      name: 'time_range_fixed',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Fix to selected Time Range'),
        description: (0, _core.t)('Fix the trend line to the full time range specified in case filtered results do not include the start or end dates'),
        renderTrigger: true,

        visibility(props) {
          const {
            time_range: timeRange
          } = props.form_data; // only display this option when a time range is selected

          return !!timeRange && timeRange !== 'No filter';
        }

      }
    }]]
  }, {
    label: (0, _core.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_picker', null], [_sharedControls.headerFontSize], [_sharedControls.subheaderFontSize]]
  }, {
    label: (0, _core.t)('Advanced Analytics'),
    expanded: false,
    controlSetRows: [// eslint-disable-next-line react/jsx-key
    [(0, _react2.jsx)("h1", {
      className: "section-header"
    }, (0, _core.t)('Rolling Window'))], [{
      name: 'rolling_type',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Rolling Function'),
        default: 'None',
        choices: (0, _chartControls.formatSelectOptions)(['None', 'mean', 'sum', 'std', 'cumsum']),
        description: (0, _core.t)('Defines a rolling window function to apply, works along ' + 'with the [Periods] text box')
      }
    }], [{
      name: 'rolling_periods',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Periods'),
        isInt: true,
        description: (0, _core.t)('Defines the size of the rolling window function, ' + 'relative to the time granularity selected')
      }
    }], [{
      name: 'min_periods',
      config: {
        type: 'TextControl',
        label: (0, _core.t)('Min Periods'),
        isInt: true,
        description: (0, _core.t)('The minimum number of rolling periods required to show ' + 'a value. For instance if you do a cumulative sum on 7 days ' + 'you may want your "Min Period" to be 7, so that all data points ' + 'shown are the total of 7 periods. This will hide the "ramp up" ' + 'taking place over the first 7 periods')
      }
    }]]
  }],
  controlOverrides: {
    y_axis_format: {
      label: (0, _core.t)('Number format')
    }
  }
};
var _default = config;
exports.default = _default;
