// react-json-tree/lib/cjs/JSONNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _objType = _interopRequireDefault(require("./objType"));

var _JSONObjectNode = _interopRequireDefault(require("./JSONObjectNode"));

var _JSONArrayNode = _interopRequireDefault(require("./JSONArrayNode"));

var _JSONIterableNode = _interopRequireDefault(require("./JSONIterableNode"));

var _JSONValueNode = _interopRequireDefault(require("./JSONValueNode"));

var _excluded = ["getItemString", "keyPath", "labelRenderer", "styling", "value", "valueRenderer", "isCustomNode"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var JSONNode = function JSONNode(_ref) {
  var getItemString = _ref.getItemString,
      keyPath = _ref.keyPath,
      labelRenderer = _ref.labelRenderer,
      styling = _ref.styling,
      value = _ref.value,
      valueRenderer = _ref.valueRenderer,
      isCustomNode = _ref.isCustomNode,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var nodeType = isCustomNode(value) ? 'Custom' : (0, _objType.default)(value);
  var simpleNodeProps = {
    getItemString: getItemString,
    key: keyPath[0],
    keyPath: keyPath,
    labelRenderer: labelRenderer,
    nodeType: nodeType,
    styling: styling,
    value: value,
    valueRenderer: valueRenderer
  };

  var nestedNodeProps = _objectSpread(_objectSpread(_objectSpread({}, rest), simpleNodeProps), {}, {
    data: value,
    isCustomNode: isCustomNode
  });

  switch (nodeType) {
    case 'Object':
    case 'Error':
    case 'WeakMap':
    case 'WeakSet':
      return /*#__PURE__*/_react.default.createElement(_JSONObjectNode.default, nestedNodeProps);

    case 'Array':
      return /*#__PURE__*/_react.default.createElement(_JSONArrayNode.default, nestedNodeProps);

    case 'Iterable':
    case 'Map':
    case 'Set':
      return /*#__PURE__*/_react.default.createElement(_JSONIterableNode.default, nestedNodeProps);

    case 'String':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter(raw) {
          return "\"".concat(raw, "\"");
        }
      }));

    case 'Number':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, simpleNodeProps);

    case 'Boolean':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter(raw) {
          return raw ? 'true' : 'false';
        }
      }));

    case 'Date':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter(raw) {
          return raw.toISOString();
        }
      }));

    case 'Null':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter() {
          return 'null';
        }
      }));

    case 'Undefined':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter() {
          return 'undefined';
        }
      }));

    case 'Function':
    case 'Symbol':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter(raw) {
          return raw.toString();
        }
      }));

    case 'Custom':
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, simpleNodeProps);

    default:
      return /*#__PURE__*/_react.default.createElement(_JSONValueNode.default, (0, _extends2.default)({}, simpleNodeProps, {
        valueGetter: function valueGetter() {
          return "<".concat(nodeType, ">");
        }
      }));
  }
};

JSONNode.propTypes = {
  getItemString: _propTypes.default.func.isRequired,
  keyPath: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired).isRequired,
  labelRenderer: _propTypes.default.func.isRequired,
  styling: _propTypes.default.func.isRequired,
  value: _propTypes.default.any,
  valueRenderer: _propTypes.default.func.isRequired,
  isCustomNode: _propTypes.default.func.isRequired
};
var _default = JSONNode;
exports.default = _default;
