// @applitools/ufg-client/dist/resources/fetch-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeFetchResource = void 0;
const req_1 = require("@applitools/req");
const resource_1 = require("./resource");
const create_cookie_header_1 = require("../utils/create-cookie-header");
const create_user_agent_header_1 = require("../utils/create-user-agent-header");
function makeFetchResource({ retryLimit = 5, streamingTimeout = 30 * 1000, cache = new Map(), fetch, logger, } = {}) {
    const req = (0, req_1.makeReq)({
        retry: {
            limit: retryLimit,
            validate: ({ error }) => Boolean(error),
        },
        fetch,
    });
    return async function fetchResource({ resource, settings = {}, }) {
        var _a;
        let runningRequest = cache.get(resource.id);
        if (runningRequest)
            return runningRequest;
        runningRequest = req(resource.url, {
            headers: {
                Referer: settings.referer,
                Cookie: (0, create_cookie_header_1.createCookieHeader)({ url: resource.url, cookies: settings.cookies }),
                'User-Agent': (_a = (0, create_user_agent_header_1.createUserAgentHeader)({ renderer: resource.renderer })) !== null && _a !== void 0 ? _a : settings.userAgent,
            },
            proxy: resourceUrl => {
                const { proxy, autProxy } = settings;
                if (autProxy) {
                    if (!autProxy.domains)
                        return autProxy;
                    const domainMatch = autProxy.domains.includes(resourceUrl.hostname);
                    if ((autProxy.mode === 'Allow' && domainMatch) || (autProxy.mode === 'Block' && !domainMatch))
                        return autProxy;
                }
                return proxy;
            },
            hooks: [handleLogs({ logger }), handleStreaming({ timeout: streamingTimeout, logger })],
        })
            .then(async (response) => {
            return response.ok
                ? (0, resource_1.makeResource)({
                    ...resource,
                    value: Buffer.from(await response.arrayBuffer()),
                    contentType: response.headers.get('Content-Type'),
                })
                : (0, resource_1.makeResource)({ ...resource, errorStatusCode: response.status });
        })
            .finally(() => cache.delete(resource.id));
        cache.set(resource.id, runningRequest);
        return runningRequest;
    };
}
exports.makeFetchResource = makeFetchResource;
function handleLogs({ logger }) {
    return {
        beforeRequest({ request }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} will be fetched using headers`, Object.fromEntries(request.headers.entries()));
        },
        beforeRetry({ request, attempt }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} will be re-fetched (attempt ${attempt})`);
        },
        afterResponse({ request, response }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} respond with ${response.statusText}(${response.statusText})`);
        },
        afterError({ request, error }) {
            logger === null || logger === void 0 ? void 0 : logger.error(`Resource with url ${request.url} failed with error`, error);
        },
    };
}
function handleStreaming({ timeout, logger }) {
    const controller = new req_1.AbortController();
    return {
        async beforeRequest({ request }) {
            var _a, _b;
            if ((_a = request.signal) === null || _a === void 0 ? void 0 : _a.aborted)
                return;
            (_b = request.signal) === null || _b === void 0 ? void 0 : _b.addEventListener('abort', () => controller.abort(), { once: true });
            return new req_1.Request(request, { signal: controller.signal });
        },
        async afterResponse({ response }) {
            const isProbablyStreaming = response.ok &&
                !response.headers.get('Content-Length') &&
                ['audio/', 'video/'].some(prefix => response.headers.get('Content-Type').startsWith(prefix));
            if (!isProbablyStreaming)
                return;
            return new Promise(resolve => {
                const timer = setTimeout(() => {
                    controller.abort();
                    resolve(new req_1.Response(null, { status: 599 }));
                    logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${response.url} was interrupted, due to it takes too long to download`);
                }, timeout);
                response
                    .arrayBuffer()
                    .then(body => resolve(new req_1.Response(body, response)))
                    .catch(() => resolve(new req_1.Response(null, { status: 599 })))
                    .finally(() => clearTimeout(timer));
            });
        },
    };
}
